{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8fed2757",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "from Bio.SeqUtils import gc_fraction\n",
    "from Bio.Seq import Seq\n",
    "from itertools import product\n",
    "from collections import Counter\n",
    "import pandas as pd\n",
    "import os\n",
    "from pathlib import Path\n",
    "from BCBio import GFF\n",
    "import re\n",
    "import itertools\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8548bd55",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ðŸ” Processing: Brandao_MCCM_full_raw_counts_tpm_filtered_classified.tsv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lelo3/anaconda3/envs/vividVirions/lib/python3.13/site-packages/Bio/Entrez/__init__.py:734: UserWarning: \n",
      "            Email address is not specified.\n",
      "\n",
      "            To make use of NCBI's E-utilities, NCBI requires you to specify your\n",
      "            email address with each request.  As an example, if your email address\n",
      "            is A.N.Other@example.com, you can specify it as follows:\n",
      "               from Bio import Entrez\n",
      "               Entrez.email = 'A.N.Other@example.com'\n",
      "            In case of excessive usage of the E-utilities, NCBI will attempt to contact\n",
      "            a user at the email address provided before blocking access to the\n",
      "            E-utilities.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[2]\u001b[39m\u001b[32m, line 62\u001b[39m\n\u001b[32m     59\u001b[39m dna_seqs = []\n\u001b[32m     61\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m geneid \u001b[38;5;129;01min\u001b[39;00m df[\u001b[33m\"\u001b[39m\u001b[33mGeneid\u001b[39m\u001b[33m\"\u001b[39m]:\n\u001b[32m---> \u001b[39m\u001b[32m62\u001b[39m     protein, dna = get_sequences_from_geneid(genome_acc, geneid)\n\u001b[32m     63\u001b[39m     protein_seqs.append(protein)\n\u001b[32m     64\u001b[39m     dna_seqs.append(dna)\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[2]\u001b[39m\u001b[32m, line 21\u001b[39m, in \u001b[36mget_sequences_from_geneid\u001b[39m\u001b[34m(genome_accession, geneid)\u001b[39m\n\u001b[32m     19\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mget_sequences_from_geneid\u001b[39m(genome_accession, geneid):\n\u001b[32m     20\u001b[39m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m---> \u001b[39m\u001b[32m21\u001b[39m         handle = Entrez.efetch(db=\u001b[33m\"\u001b[39m\u001b[33mnucleotide\u001b[39m\u001b[33m\"\u001b[39m, \u001b[38;5;28mid\u001b[39m=genome_accession, rettype=\u001b[33m\"\u001b[39m\u001b[33mgb\u001b[39m\u001b[33m\"\u001b[39m, retmode=\u001b[33m\"\u001b[39m\u001b[33mtext\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m     22\u001b[39m         record = SeqIO.read(handle, \u001b[33m\"\u001b[39m\u001b[33mgenbank\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m     23\u001b[39m         handle.close()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/vividVirions/lib/python3.13/site-packages/Bio/Entrez/__init__.py:201\u001b[39m, in \u001b[36mefetch\u001b[39m\u001b[34m(db, **keywords)\u001b[39m\n\u001b[32m    199\u001b[39m variables.update(keywords)\n\u001b[32m    200\u001b[39m request = _build_request(cgi, variables)\n\u001b[32m--> \u001b[39m\u001b[32m201\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m _open(request)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/vividVirions/lib/python3.13/site-packages/Bio/Entrez/__init__.py:634\u001b[39m, in \u001b[36m_open\u001b[39m\u001b[34m(request)\u001b[39m\n\u001b[32m    632\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(max_tries):\n\u001b[32m    633\u001b[39m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m634\u001b[39m         handle = urlopen(request)\n\u001b[32m    635\u001b[39m     \u001b[38;5;28;01mexcept\u001b[39;00m HTTPError \u001b[38;5;28;01mas\u001b[39;00m exception:\n\u001b[32m    636\u001b[39m         \u001b[38;5;66;03m# Reraise if the final try fails\u001b[39;00m\n\u001b[32m    637\u001b[39m         \u001b[38;5;28;01mif\u001b[39;00m i >= max_tries - \u001b[32m1\u001b[39m:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/vividVirions/lib/python3.13/urllib/request.py:189\u001b[39m, in \u001b[36murlopen\u001b[39m\u001b[34m(url, data, timeout, context)\u001b[39m\n\u001b[32m    187\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    188\u001b[39m     opener = _opener\n\u001b[32m--> \u001b[39m\u001b[32m189\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m opener.open(url, data, timeout)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/vividVirions/lib/python3.13/urllib/request.py:489\u001b[39m, in \u001b[36mOpenerDirector.open\u001b[39m\u001b[34m(self, fullurl, data, timeout)\u001b[39m\n\u001b[32m    486\u001b[39m     req = meth(req)\n\u001b[32m    488\u001b[39m sys.audit(\u001b[33m'\u001b[39m\u001b[33murllib.Request\u001b[39m\u001b[33m'\u001b[39m, req.full_url, req.data, req.headers, req.get_method())\n\u001b[32m--> \u001b[39m\u001b[32m489\u001b[39m response = \u001b[38;5;28mself\u001b[39m._open(req, data)\n\u001b[32m    491\u001b[39m \u001b[38;5;66;03m# post-process response\u001b[39;00m\n\u001b[32m    492\u001b[39m meth_name = protocol+\u001b[33m\"\u001b[39m\u001b[33m_response\u001b[39m\u001b[33m\"\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/vividVirions/lib/python3.13/urllib/request.py:506\u001b[39m, in \u001b[36mOpenerDirector._open\u001b[39m\u001b[34m(self, req, data)\u001b[39m\n\u001b[32m    503\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n\u001b[32m    505\u001b[39m protocol = req.type\n\u001b[32m--> \u001b[39m\u001b[32m506\u001b[39m result = \u001b[38;5;28mself\u001b[39m._call_chain(\u001b[38;5;28mself\u001b[39m.handle_open, protocol, protocol +\n\u001b[32m    507\u001b[39m                           \u001b[33m'\u001b[39m\u001b[33m_open\u001b[39m\u001b[33m'\u001b[39m, req)\n\u001b[32m    508\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m result:\n\u001b[32m    509\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/vividVirions/lib/python3.13/urllib/request.py:466\u001b[39m, in \u001b[36mOpenerDirector._call_chain\u001b[39m\u001b[34m(self, chain, kind, meth_name, *args)\u001b[39m\n\u001b[32m    464\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m handler \u001b[38;5;129;01min\u001b[39;00m handlers:\n\u001b[32m    465\u001b[39m     func = \u001b[38;5;28mgetattr\u001b[39m(handler, meth_name)\n\u001b[32m--> \u001b[39m\u001b[32m466\u001b[39m     result = func(*args)\n\u001b[32m    467\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m result \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m    468\u001b[39m         \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/vividVirions/lib/python3.13/urllib/request.py:1367\u001b[39m, in \u001b[36mHTTPSHandler.https_open\u001b[39m\u001b[34m(self, req)\u001b[39m\n\u001b[32m   1366\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mhttps_open\u001b[39m(\u001b[38;5;28mself\u001b[39m, req):\n\u001b[32m-> \u001b[39m\u001b[32m1367\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m.do_open(http.client.HTTPSConnection, req,\n\u001b[32m   1368\u001b[39m                         context=\u001b[38;5;28mself\u001b[39m._context)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/vividVirions/lib/python3.13/urllib/request.py:1323\u001b[39m, in \u001b[36mAbstractHTTPHandler.do_open\u001b[39m\u001b[34m(self, http_class, req, **http_conn_args)\u001b[39m\n\u001b[32m   1321\u001b[39m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err: \u001b[38;5;66;03m# timeout error\u001b[39;00m\n\u001b[32m   1322\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m URLError(err)\n\u001b[32m-> \u001b[39m\u001b[32m1323\u001b[39m     r = h.getresponse()\n\u001b[32m   1324\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m:\n\u001b[32m   1325\u001b[39m     h.close()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/vividVirions/lib/python3.13/http/client.py:1430\u001b[39m, in \u001b[36mHTTPConnection.getresponse\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m   1428\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m   1429\u001b[39m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m-> \u001b[39m\u001b[32m1430\u001b[39m         response.begin()\n\u001b[32m   1431\u001b[39m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m:\n\u001b[32m   1432\u001b[39m         \u001b[38;5;28mself\u001b[39m.close()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/vividVirions/lib/python3.13/http/client.py:331\u001b[39m, in \u001b[36mHTTPResponse.begin\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m    329\u001b[39m \u001b[38;5;66;03m# read until we get a non-100 response\u001b[39;00m\n\u001b[32m    330\u001b[39m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m331\u001b[39m     version, status, reason = \u001b[38;5;28mself\u001b[39m._read_status()\n\u001b[32m    332\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m status != CONTINUE:\n\u001b[32m    333\u001b[39m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/vividVirions/lib/python3.13/http/client.py:292\u001b[39m, in \u001b[36mHTTPResponse._read_status\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m    291\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34m_read_status\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[32m--> \u001b[39m\u001b[32m292\u001b[39m     line = \u001b[38;5;28mstr\u001b[39m(\u001b[38;5;28mself\u001b[39m.fp.readline(_MAXLINE + \u001b[32m1\u001b[39m), \u001b[33m\"\u001b[39m\u001b[33miso-8859-1\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m    293\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(line) > _MAXLINE:\n\u001b[32m    294\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m LineTooLong(\u001b[33m\"\u001b[39m\u001b[33mstatus line\u001b[39m\u001b[33m\"\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/vividVirions/lib/python3.13/socket.py:719\u001b[39m, in \u001b[36mSocketIO.readinto\u001b[39m\u001b[34m(self, b)\u001b[39m\n\u001b[32m    717\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m(\u001b[33m\"\u001b[39m\u001b[33mcannot read from timed out object\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m    718\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m719\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m._sock.recv_into(b)\n\u001b[32m    720\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m timeout:\n\u001b[32m    721\u001b[39m     \u001b[38;5;28mself\u001b[39m._timeout_occurred = \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/vividVirions/lib/python3.13/ssl.py:1304\u001b[39m, in \u001b[36mSSLSocket.recv_into\u001b[39m\u001b[34m(self, buffer, nbytes, flags)\u001b[39m\n\u001b[32m   1300\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m flags != \u001b[32m0\u001b[39m:\n\u001b[32m   1301\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[32m   1302\u001b[39m           \u001b[33m\"\u001b[39m\u001b[33mnon-zero flags not allowed in calls to recv_into() on \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[33m\"\u001b[39m %\n\u001b[32m   1303\u001b[39m           \u001b[38;5;28mself\u001b[39m.\u001b[34m__class__\u001b[39m)\n\u001b[32m-> \u001b[39m\u001b[32m1304\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m.read(nbytes, buffer)\n\u001b[32m   1305\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m   1306\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m().recv_into(buffer, nbytes, flags)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/vividVirions/lib/python3.13/ssl.py:1138\u001b[39m, in \u001b[36mSSLSocket.read\u001b[39m\u001b[34m(self, len, buffer)\u001b[39m\n\u001b[32m   1136\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m   1137\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m buffer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m-> \u001b[39m\u001b[32m1138\u001b[39m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m._sslobj.read(\u001b[38;5;28mlen\u001b[39m, buffer)\n\u001b[32m   1139\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m   1140\u001b[39m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m._sslobj.read(\u001b[38;5;28mlen\u001b[39m)\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: "
     ]
    }
   ],
   "source": [
    "input_folder = \"../data/classified\"\n",
    "output_folder = \"../data/feature_extraction\"\n",
    "\n",
    "# Map each input file to its matching genome accession\n",
    "genome_mapping = {\n",
    "    \"Brandao_MCCM_full_raw_counts_tpm_filtered_classified.tsv\": \"NC_010326\",\n",
    "    \"Finstrlova_Newman_full_raw_counts_tpm_filtered_classified.tsv\": \"NC_005880\",\n",
    "    \"Guegler_T4_minusToxIN_full_raw_counts_tpm_filtered_classified.tsv\": \"NC_000866\",\n",
    "    \"Guegler_T7_plusToxIN_full_raw_counts_tpm_filtered_classified.tsv\": \"NC_001604\",\n",
    "    \"Lood_full_raw_counts_tpm_filtered_classified.tsv\": \"MK797984.1\",\n",
    "    \"Sprenger_VC_WT_VP882_delta_cpdS_full_raw_counts_tpm_filtered_classified.tsv\": \"NC_009016.1\",\n",
    "    \"Yang_full_raw_counts_tpm_filtered_classified.tsv\": \"NC_021316\",\n",
    "}\n",
    "\n",
    "# Make sure output folder exists\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "\n",
    "# ---- FUNCTION TO GET PROTEIN + DNA SEQUENCE ----\n",
    "def get_sequences_from_geneid(genome_accession, geneid):\n",
    "    try:\n",
    "        handle = Entrez.efetch(\n",
    "            db=\"nucleotide\", id=genome_accession, rettype=\"gb\", retmode=\"text\"\n",
    "        )\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "        handle.close()\n",
    "    except Exception as e:\n",
    "        print(f\"âš  Failed to fetch {genome_accession}: {e}\")\n",
    "        return (\"ERROR_FETCH\", \"ERROR_FETCH\")\n",
    "\n",
    "    tag = geneid.replace(\"gene-\", \"\").strip()\n",
    "    short_tag = tag.split(\"_\")[-1]\n",
    "\n",
    "    for feature in record.features:\n",
    "        if feature.type == \"CDS\":\n",
    "            locus_tag = feature.qualifiers.get(\"locus_tag\", [\"\"])[0]\n",
    "            gene = feature.qualifiers.get(\"gene\", [\"\"])[0]\n",
    "            product = feature.qualifiers.get(\"product\", [\"\"])[0]\n",
    "\n",
    "            if tag in [locus_tag, gene, product] or short_tag in [\n",
    "                locus_tag,\n",
    "                gene,\n",
    "                product,\n",
    "            ]:\n",
    "                protein = feature.qualifiers.get(\n",
    "                    \"translation\", [\"TRANSLATION_NOT_FOUND\"]\n",
    "                )[0]\n",
    "                dna_seq = feature.location.extract(record.seq)\n",
    "                return (protein, str(dna_seq))\n",
    "\n",
    "    return (\"NOT_FOUND\", \"NOT_FOUND\")\n",
    "\n",
    "\n",
    "# ---- MAIN PROCESSING ----\n",
    "for file_name in os.listdir(input_folder):\n",
    "    if not file_name.endswith(\".tsv\"):\n",
    "        continue\n",
    "\n",
    "    print(f\"ðŸ” Processing: {file_name}\")\n",
    "    file_path = os.path.join(input_folder, file_name)\n",
    "    df = pd.read_csv(file_path, sep=\"\\t\")\n",
    "\n",
    "    genome_acc = genome_mapping.get(file_name)\n",
    "    if not genome_acc:\n",
    "        print(f\"âš  No genome accession mapped for {file_name}\")\n",
    "        continue\n",
    "\n",
    "    protein_seqs = []\n",
    "    dna_seqs = []\n",
    "\n",
    "    for geneid in df[\"Geneid\"]:\n",
    "        protein, dna = get_sequences_from_geneid(genome_acc, geneid)\n",
    "        protein_seqs.append(protein)\n",
    "        dna_seqs.append(dna)\n",
    "\n",
    "    df[\"ProteinSequence\"] = protein_seqs\n",
    "    df[\"DNASequence\"] = dna_seqs\n",
    "\n",
    "    out_path = os.path.join(output_folder, file_name)\n",
    "    df.to_csv(out_path, sep=\"\\t\", index=False)\n",
    "    print(f\"âœ… Saved to: {out_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e7b6e78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "âŒ Failed sequence extractions found in the following files:\n",
      "\n",
      "ðŸ“„ Finstrlova_Newman_full_raw_counts_tpm_filtered_classified.tsv â€” 4 failures\n",
      "               Geneid ProteinSequence DNASequence\n",
      "gene-CPT_phageK_gt004       NOT_FOUND   NOT_FOUND\n",
      "gene-CPT_phageK_gt002       NOT_FOUND   NOT_FOUND\n",
      "gene-CPT_phageK_gt003       NOT_FOUND   NOT_FOUND\n",
      "gene-CPT_phageK_gt001       NOT_FOUND   NOT_FOUND\n",
      "------------------------------------------------------------\n",
      "ðŸ“„ Guegler_T4_minusToxIN_full_raw_counts_tpm_filtered_classified.tsv â€” 10 failures\n",
      "     Geneid ProteinSequence DNASequence\n",
      "gene-T4t006       NOT_FOUND   NOT_FOUND\n",
      "gene-T4t003       NOT_FOUND   NOT_FOUND\n",
      "gene-T4t008       NOT_FOUND   NOT_FOUND\n",
      "gene-T4t007       NOT_FOUND   NOT_FOUND\n",
      "gene-T4s002       NOT_FOUND   NOT_FOUND\n",
      "gene-T4t001       NOT_FOUND   NOT_FOUND\n",
      "gene-T4s001       NOT_FOUND   NOT_FOUND\n",
      "gene-T4t002       NOT_FOUND   NOT_FOUND\n",
      "gene-T4t004       NOT_FOUND   NOT_FOUND\n",
      "gene-T4t005       NOT_FOUND   NOT_FOUND\n",
      "------------------------------------------------------------\n",
      "ðŸ“„ Lood_full_raw_counts_tpm_filtered_classified.tsv â€” 16 failures\n",
      "         Geneid ProteinSequence DNASequence\n",
      "gene-EST35_0396       NOT_FOUND   NOT_FOUND\n",
      "gene-EST35_0036       NOT_FOUND   NOT_FOUND\n",
      "gene-EST35_0406       NOT_FOUND   NOT_FOUND\n",
      "gene-EST35_0394       NOT_FOUND   NOT_FOUND\n",
      "gene-EST35_0398       NOT_FOUND   NOT_FOUND\n",
      "gene-EST35_0393       NOT_FOUND   NOT_FOUND\n",
      "gene-EST35_0353       NOT_FOUND   NOT_FOUND\n",
      "gene-EST35_0397       NOT_FOUND   NOT_FOUND\n",
      "gene-EST35_0414       NOT_FOUND   NOT_FOUND\n",
      "gene-EST35_0413       NOT_FOUND   NOT_FOUND\n",
      "gene-EST35_0407       NOT_FOUND   NOT_FOUND\n",
      "gene-EST35_0395       NOT_FOUND   NOT_FOUND\n",
      "gene-EST35_0068       NOT_FOUND   NOT_FOUND\n",
      "gene-EST35_0403       NOT_FOUND   NOT_FOUND\n",
      "gene-EST35_0408       NOT_FOUND   NOT_FOUND\n",
      "gene-EST35_0168       NOT_FOUND   NOT_FOUND\n",
      "------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# Folder containing your result files\n",
    "results_folder = \"../data/feature_extraction\"\n",
    "\n",
    "# Keywords to look for in failed extractions\n",
    "failure_keywords = [\"NOT_FOUND\", \"ERROR_FETCH\", \"TRANSLATION_NOT_FOUND\"]\n",
    "\n",
    "# Storage for summary\n",
    "summary = {}\n",
    "\n",
    "# Loop through all TSV files\n",
    "for file_name in os.listdir(results_folder):\n",
    "    if not file_name.endswith(\".tsv\"):\n",
    "        continue\n",
    "\n",
    "    file_path = os.path.join(results_folder, file_name)\n",
    "    df = pd.read_csv(file_path, sep=\"\\t\")\n",
    "\n",
    "    failed_rows = df[\n",
    "        df[\"ProteinSequence\"].isin(failure_keywords)\n",
    "        | df[\"DNASequence\"].isin(failure_keywords)\n",
    "    ]\n",
    "\n",
    "    if not failed_rows.empty:\n",
    "        summary[file_name] = failed_rows[[\"Geneid\", \"ProteinSequence\", \"DNASequence\"]]\n",
    "\n",
    "# Report the results\n",
    "if summary:\n",
    "    print(\"âŒ Failed sequence extractions found in the following files:\\n\")\n",
    "    for fname, failures in summary.items():\n",
    "        print(f\"ðŸ“„ {fname} â€” {len(failures)} failures\")\n",
    "        print(failures.to_string(index=False))\n",
    "        print(\"-\" * 60)\n",
    "else:\n",
    "    print(\"âœ… All sequence extractions succeeded. No issues found.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec687d5e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "047688c9",
   "metadata": {},
   "source": [
    "# User-story 11: Sequence based features\n",
    "@LuiseJedlitschka  \n",
    "@milli2908  \n",
    "@elivic734    \n",
    "    \n",
    "## Potential features:\n",
    "- **CG-count**: Despite being a simple metric, GC presents a huge variation across genomes (ranging from approximately 20% to 70% ). GC content is reasonably constant within a given genome, and was already found to be correlated with several universal factors of microbial lifestyles such as temperature, niche complexity and aerobiosis [Genomic Signature](https://www.sciencedirect.com/topics/biochemistry-genetics-and-molecular-biology/genomic-signature).\n",
    "- **sequence-length**\n",
    "- **K-Mer-frequency**: e.g. k3 -> codon composition\n",
    "- **nucleotide composition** (Base percentages and purin/pyrimidin percentages)\n",
    "- **CpG bias**: relevant for interaction with host, methylation, epigenetic regulation, \n",
    " Interpretation:\n",
    "\n",
    "    CpG < 1: CpG is underrepresented\n",
    "\n",
    "    CpG â‰ˆ 1: expected frequency, no bias\n",
    "\n",
    "    CpG > 1: CpG is overrepresented\n",
    "- **motives**:\n",
    "\n",
    "    \"TATA_box\": \"TATAAA\",       # Promotor\n",
    "\n",
    "    \"CAAT_box\": \"GGCCAATCT\",    # Enhancer\n",
    "\n",
    "    \"PolyA_signal\": \"AATAAA\",   # Signal for transcriptation\n",
    "- **relative position in genome**\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "## User-story 13: integration of external insights\n",
    "@LuiseJedlitschka  \n",
    "@milli2908  \n",
    "@elivic734    \n",
    "  \n",
    "[Genomic features of E. coli O177 phages](https://www.nature.com/articles/s41598-023-48788-w/tables/1): sequence length, GC-content\n",
    "\n",
    "*comparison*: also used as features in our model\n",
    "\n",
    "[DeepPL: A deep-learning-based tool for the prediction of bacteriophage lifecycle](https://pmc.ncbi.nlm.nih.gov/articles/PMC11521287/): K-mer 6 was selected in this study based on its best performance, as reported by DNABERT\n",
    "\n",
    "*comparison*: the sliding window approach is used to detect the Tricolons, as our input is not a whole genome but already specific geneso used in our model\n",
    "\n",
    "[Digital phagograms: predicting phage infectivity through a multilayer machine learning approach](https://www.sciencedirect.com/science/article/pii/S1879625721001620): \n",
    "Nucleotide composition, Codon composition, Codon usage bias (Codon usage bias refers to the differences in the frequency of usage for a certain base triplet for a given amino acid among organisms), GC content, CpG bias, k-mer frequeR spacers, Embeddings, pVOG hits  Coding DNA sequences\t(Auxiliary metabolic genes, Shared tRNAs, defense system genes, Physicochemical properties)\n",
    "\n",
    "*comparison*: only a few of these (nucleotide composition, Codon composition, GC content, k-mer frequencies, CpG bias) are used in our model as some are redundant with the structural features relying (pVOG hits, Coding DNA sequences, Physicochemical properties) on the protein sequence, some are only necessary when the and the host-genome (CRISPR spacers, Shared tRNAs, defense system genes) input is a whole genome, not just sequences like in our case, and some are just too specific considering our main focus is on the protein-sequences-features (Codon usage bias)\n",
    "\n",
    "[PhageAI - Bacteriophage Life Cycle Recognition with Machine Learning and Natural Language Processing](https://www.biorxiv.org/content/10.1101/2020.07.11.198606v1.full): sliding window approach using constant k = 6 and the Word2Vec algorithm with the Skip-gram model, nominal features were empirically chosen by feature selection algorithm called Feature ranking with recursive feature elimination and cross-validated selection of the best number of features using Support Vector Machine \n",
    "\n",
    "*comparison*: sliding window approach is used to detect the Tricolons, as our input is not a whole genome but already specific genes\n",
    "\n",
    "[Rapid discovery of novel prophages using biological feature engineering and machine learning](https://academic.oup.com/nargab/article/3/1/lqaa109/6066536?login=true): gc_content (also in specific positions of the codons), CAI (codon adaptation index, measures the codon usage bias), BP:percent (for all bases, purin, pyrimidin)\n",
    "\n",
    "*comparison*: the most important (gc_content and BP:percentage) are in our model, the others are very detailed and unnecessary as our features focus on protein sequenceson: also used in our model\n",
    "                                                                                                        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f60a48f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Features table saved as ../data/dna_feature_table\n",
      "          Geneid                                        DNASequence  \\\n",
      "0    gene-T4p066  ATGAACGAAGCTCTTATTAACGATTTACGTCTCGCTGGGTATGAAG...   \n",
      "1    gene-T4p122  ATGAAAACATATCAAGAATTTATTGCCGAAGCTTCTGTAGTAAAGG...   \n",
      "2    gene-T4p143  ATGCTCTTTACATTTTTTGATCCGATTGAATATGCGGCCAAAACGG...   \n",
      "3    gene-T4p252  ATGGCTACTTTAAAACAAATACAATTTAAAAGAAGCAAAATCGCAG...   \n",
      "4    gene-T4p127  ATGTCAAAGCCATCACTATACCTTCCAAGTAAACCTTTGAAGTATG...   \n",
      "..           ...                                                ...   \n",
      "283  gene-T4p187  ATGAGATTAGAAGATCTTCAAGAAGAATTGAAGAAAGATGTGTTTA...   \n",
      "284  gene-T4p078  ATGTTATTGACTGGCAAATTATACAAAGAAGAAAAACAGAAATTTT...   \n",
      "285  gene-T4p263  ATGAGTCCATTCATTGGTATTACAAGCGCTGCATTAGTATCTGGTA...   \n",
      "286  gene-T4p191  ATGGCAAACATTATTCGTTGTAAATTACCAGATGGTGTTCATCGTT...   \n",
      "287  gene-T4p150  ATGTCAGGATTAAGTTATGATAAGTGTGTTACTGCTGGCCATGAAG...   \n",
      "\n",
      "    classification Gene_Position_Relative  GC_Content  Seq_length  A_Content  \\\n",
      "0            early                   None    0.356061         264   0.310606   \n",
      "1            early                   None    0.407216         582   0.312715   \n",
      "2             late                   None    0.340102         591   0.363790   \n",
      "3             late                   None    0.409283        3081   0.297306   \n",
      "4            early                   None    0.333333         393   0.340967   \n",
      "..             ...                    ...         ...         ...        ...   \n",
      "283           late                   None    0.345411         414   0.357488   \n",
      "284          early                   None    0.337553         474   0.386076   \n",
      "285           late                   None    0.446667         150   0.226667   \n",
      "286           late                   None    0.298667         750   0.386667   \n",
      "287           late                   None    0.408163         294   0.333333   \n",
      "\n",
      "     T_Content  G_Content  C_Content  ...  kmer_CTG  kmer_CTC  kmer_CGA  \\\n",
      "0     0.333333   0.208333   0.147727  ...  0.007634  0.019084  0.011450   \n",
      "1     0.280069   0.218213   0.189003  ...  0.010345  0.008621  0.010345   \n",
      "2     0.296108   0.186125   0.153976  ...  0.008489  0.005093  0.010187   \n",
      "3     0.293411   0.224927   0.184356  ...  0.017863  0.009419  0.007795   \n",
      "4     0.325700   0.188295   0.145038  ...  0.015345  0.002558  0.002558   \n",
      "..         ...        ...        ...  ...       ...       ...       ...   \n",
      "283   0.297101   0.234300   0.111111  ...  0.004854  0.009709  0.009709   \n",
      "284   0.276371   0.187764   0.149789  ...  0.010593  0.004237  0.008475   \n",
      "285   0.326667   0.253333   0.193333  ...  0.020270  0.000000  0.000000   \n",
      "286   0.314667   0.166667   0.132000  ...  0.010695  0.004011  0.010695   \n",
      "287   0.258503   0.227891   0.180272  ...  0.027397  0.010274  0.003425   \n",
      "\n",
      "     kmer_CGT  kmer_CGG  kmer_CGC  kmer_CCA  kmer_CCT  kmer_CCG  kmer_CCC  \n",
      "0    0.003817  0.007634  0.007634  0.007634  0.007634  0.000000  0.000000  \n",
      "1    0.006897  0.008621  0.013793  0.003448  0.012069  0.018966  0.005172  \n",
      "2    0.010187  0.010187  0.013582  0.008489  0.005093  0.008489  0.003396  \n",
      "3    0.009094  0.012666  0.006820  0.011367  0.005521  0.007145  0.003897  \n",
      "4    0.012788  0.000000  0.005115  0.020460  0.007673  0.002558  0.005115  \n",
      "..        ...       ...       ...       ...       ...       ...       ...  \n",
      "283  0.007282  0.007282  0.000000  0.000000  0.002427  0.000000  0.000000  \n",
      "284  0.006356  0.002119  0.000000  0.012712  0.006356  0.004237  0.004237  \n",
      "285  0.027027  0.013514  0.006757  0.013514  0.000000  0.006757  0.000000  \n",
      "286  0.006684  0.005348  0.002674  0.012032  0.005348  0.002674  0.001337  \n",
      "287  0.010274  0.006849  0.000000  0.013699  0.006849  0.003425  0.000000  \n",
      "\n",
      "[288 rows x 80 columns]\n",
      "Features table saved as ../data/dna_feature_table\n",
      "              Geneid                                        DNASequence  \\\n",
      "0    gene-EST35_0151  ATGTCTCATCGTAAGGAACTAGAAAAAGTTCTAGATCTTCTACTAA...   \n",
      "1    gene-EST35_0087  ATGGCTAAAGTATTGCATGTTGACAGCATCGGCAACATCGCCGAAA...   \n",
      "2    gene-EST35_0411  ATGGAAAACATATTTAGGATATCTATTAATGCTAACGAAGGTATTA...   \n",
      "3    gene-EST35_0027  ATGTTAGATATATTGCTACGGTTCCAACGCAGAGTGGATACAATTA...   \n",
      "4    gene-EST35_0377  ATGAAAAATATAATTAAACATATTTTAGATTTTTATAACGGATTTT...   \n",
      "..               ...                                                ...   \n",
      "456  gene-EST35_0345  ATGGCTAAGAAAGCTAAAGCTGTAGAAAAAGAAGTTGTAGAAAAAG...   \n",
      "457  gene-EST35_0288  ATGAACGACAAATATAATAGAAAAGAAATAACTATATCAACAGGCA...   \n",
      "458  gene-EST35_0242  ATGACAGATTGGAATAATCTCCGGGCAGATAAACGGATTGCTGCCT...   \n",
      "459  gene-EST35_0168                                          NOT_FOUND   \n",
      "460  gene-EST35_0455  ATGTCAATTCATTACTTAATACAATGTTATGTAAAGAATATTATGA...   \n",
      "\n",
      "    classification  Gene_Position_Relative  GC_Content  Seq_length  A_Content  \\\n",
      "0             late                0.291694    0.344403         813   0.399754   \n",
      "1            early                0.100072    0.522088         249   0.224900   \n",
      "2           middle                0.945150    0.250000         168   0.357143   \n",
      "3           middle                0.019904    0.379085         153   0.300654   \n",
      "4           middle                0.902407    0.284848         330   0.330303   \n",
      "..             ...                     ...         ...         ...        ...   \n",
      "456           late                0.842163    0.376238         303   0.409241   \n",
      "457           late                0.654746    0.305882         510   0.380392   \n",
      "458           late                0.559207    0.314943         435   0.365517   \n",
      "459           late                0.360433    0.000000           9   0.000000   \n",
      "460         middle                0.993445    0.256881         327   0.391437   \n",
      "\n",
      "     T_Content  G_Content  C_Content  ...  kmer_CTG  kmer_CTC  kmer_CGA  \\\n",
      "0     0.255843   0.204182   0.140221  ...  0.022195  0.007398  0.009864   \n",
      "1     0.253012   0.261044   0.261044  ...  0.020243  0.016194  0.020243   \n",
      "2     0.392857   0.130952   0.119048  ...  0.000000  0.000000  0.018072   \n",
      "3     0.320261   0.222222   0.156863  ...  0.006623  0.000000  0.019868   \n",
      "4     0.384848   0.181818   0.103030  ...  0.015244  0.003049  0.003049   \n",
      "..         ...        ...        ...  ...       ...       ...       ...   \n",
      "456   0.214521   0.240924   0.135314  ...  0.013289  0.003322  0.003322   \n",
      "457   0.313725   0.178431   0.127451  ...  0.009843  0.003937  0.007874   \n",
      "458   0.319540   0.160920   0.154023  ...  0.018476  0.009238  0.002309   \n",
      "459   0.111111   0.000000   0.000000  ...  0.000000  0.000000  0.000000   \n",
      "460   0.351682   0.119266   0.137615  ...  0.009231  0.000000  0.003077   \n",
      "\n",
      "     kmer_CGT  kmer_CGG  kmer_CGC  kmer_CCA  kmer_CCT  kmer_CCG  kmer_CCC  \n",
      "0    0.007398  0.002466  0.003699  0.000000  0.011097  0.002466  0.002466  \n",
      "1    0.024291  0.012146  0.028340  0.008097  0.020243  0.016194  0.004049  \n",
      "2    0.000000  0.006024  0.000000  0.000000  0.006024  0.006024  0.018072  \n",
      "3    0.006623  0.013245  0.013245  0.006623  0.000000  0.026490  0.006623  \n",
      "4    0.006098  0.015244  0.000000  0.000000  0.003049  0.009146  0.000000  \n",
      "..        ...       ...       ...       ...       ...       ...       ...  \n",
      "456  0.006645  0.006645  0.009967  0.009967  0.006645  0.003322  0.000000  \n",
      "457  0.003937  0.005906  0.003937  0.005906  0.003937  0.001969  0.000000  \n",
      "458  0.002309  0.004619  0.000000  0.009238  0.013857  0.006928  0.004619  \n",
      "459  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
      "460  0.003077  0.000000  0.006154  0.009231  0.003077  0.003077  0.003077  \n",
      "\n",
      "[461 rows x 80 columns]\n",
      "Features table saved as ../data/dna_feature_table\n",
      "        Geneid                                        DNASequence  \\\n",
      "0   gene-T7p48  ATGATGACTATAAGACCTACTAAAAGTACAGACTTTGAGGTATTCA...   \n",
      "1   gene-T7p43  ATGGCTGAATCTAATGCAGACGTATATGCATCTTTTGGCGTGAACT...   \n",
      "2   gene-T7p49  ATGTGTTGGGCAGCCGCAATACCTATCGCTATATCTGGCGCTCAGG...   \n",
      "3   gene-T7p05  ATGATGAAGCACTACGTTATGCCAATCCACACGTCCAACGGGGCAA...   \n",
      "4   gene-T7p47  ATGGCACTCATTAGCCAATCAATCAAGAACTTGAAGGGTGGTATCA...   \n",
      "5   gene-T7p55  ATGCTAGAATTTTTACGTAAGCTAATCCCTTGGGTTCTCGCTGGGA...   \n",
      "6   gene-T7p38  ATGTGTTTCTCACCGAAAATTAAAACTCCGAAGATGGATACCAATC...   \n",
      "7   gene-T7p19  ATGGCAGGTTACGGCGCTAAAGGAATCCGAAAGGTTGGAGCGTTTC...   \n",
      "8   gene-T7p36  GTGGAGACCGTAGCGTATATTGGTCTGGGTCTTTGTGTTCTCGGAG...   \n",
      "9   gene-T7p39  ATGTCTGAGTTCACATGTGTGGAGGCTAAGAGTCGCTTCCGTGCAA...   \n",
      "10  gene-T7p52  ATGGCTAACGTAATTAAAACCGTTTTGACTTACCAGTTAGATGGCT...   \n",
      "11  gene-T7p33  ATGTCTGACTACCTGAAAGTGCTGCAAGCAATCAAAAGTTGCCCTA...   \n",
      "12  gene-T7p42  ATGGCTGAGAAACGAACAGGACTTGCGGAGGATGGCGCAAAGTCTG...   \n",
      "13  gene-T7p41  GTGGAAGACTGCATTGAATGGACCGGAGGTGTCAACTCTAAGGGTT...   \n",
      "14  gene-T7p07  ATGAACACGATTAACATCGCTAAGAACGACTTCTCTGACATCGAAC...   \n",
      "15  gene-T7p26  ATGTTCAAACTGATTAAGAAGTTAGGCCAACTGCTGGTTCGTATGT...   \n",
      "16  gene-T7p37  ATGTTAACACCTATTAACCAATTACTTAAGAACCCTAACGATATTC...   \n",
      "17  gene-T7p54  ATGGAAAAGGATAAGAGCCTTATTACATTCTTAGAGATGTTGGACA...   \n",
      "18  gene-T7p30  ATGAACGAAAGACACTTAACAGGTGCTGCTTCTGAAATGCTAGTAG...   \n",
      "19  gene-T7p58  GTGGGTACACAACCATTATCTGGCCTGCTCTGTACCCAAGGACACG...   \n",
      "20  gene-T7p32  ATGGCTATGACAAAGAAATTTAAAGTGTCCTTCGACGTTACCGCAA...   \n",
      "21  gene-T7p10  ATGATGAACATTAAGACTAACCCGTTTAAAGCCGTGTCTTTCGTAG...   \n",
      "22  gene-T7p28  ATGCGTGACCCTAAAGTTATCCAAGCAGAAATCGCTAAACTGGAAG...   \n",
      "23  gene-T7p25  ATGAGGAAGGTCGCCCCGTTTCTATTACTGACCTACGTGGTTCTGG...   \n",
      "24  gene-T7p09  ATGCGTAACTTCGAAAAGATGACCAAACGTTCTAACCGTAATGCTC...   \n",
      "25  gene-T7p13  ATGTTTCGACTTCATTACAACAAAAGCGTTAAGAATTTCACGGTTC...   \n",
      "26  gene-T7p15  ATGCATAACTTCAAGTCAACCCCACCTGCCGACAGCCTATCTGATG...   \n",
      "27  gene-T7p14  ATGGGACTGTTAGATGGTGAAGCCTGGGAAAAAGAAAACCCGCCAG...   \n",
      "28  gene-T7p53  GTGCTATCATTAGACTTTAACAACGAATTGATTAAGGCTGCTCCAA...   \n",
      "29  gene-T7p56  ATGAGTACGTTAAGAGAGTTGAGGCTGCGAAGAGCACTCAAAGAGC...   \n",
      "30  gene-T7p17  ATGGCTAAGAAGATTTTCACCTCTGCGCTGGGTACCGCTGAACCTT...   \n",
      "31  gene-T7p18  GTGGAACTGCGGGAGAAAATCCTTGAGCGAATCAAGGTGACTTCCT...   \n",
      "32  gene-T7p08  ATGGGACGTTTATATAGTGGTAATCTGGCAGCATTCAAGGCAGCAA...   \n",
      "33  gene-T7p02  ATGTCTACTACCAACGTGCAATACGGTCTGACCGCTCAAACTGTAC...   \n",
      "34  gene-T7p21  ATGCGCAAGTCTTATAAACAATTCTATAAGGCTCCGAGGAGGCATA...   \n",
      "35  gene-T7p46  ATGCGCTCATACGATATGAACGTTGAGACTGCCGCTGAGTTATCAG...   \n",
      "36  gene-T7p34  ATGTCTCGTGACCTTGTGACTATTCCACGCGATGTGTGGAACGATA...   \n",
      "37  gene-T7p27  ATGTCTAACGTAGCTGAAACTATCCGTCTATCCGATACAGCTGACC...   \n",
      "38  gene-T7p57  GTGTCTACTCAATCCAATCGTAATGCGCTCGTAGTGGCGCAACTGA...   \n",
      "39  gene-T7p22  ATGGACAATTCGCACGATTCCGATAGTGTATTTCTTTACCACATTC...   \n",
      "40  gene-T7p51  ATGGATAAGTACGATAAGAACGTACCAAGTGATTATGATGGTCTGT...   \n",
      "41  gene-T7p20  ATGGCTCGTGTACAGTTTAAACAACGTGAATCTACTGACGCAATCT...   \n",
      "42  gene-T7p03  ATGAACATTACCGACATCATGAACGCTATCGACGCAATCAAAGCAC...   \n",
      "43  gene-T7p40  ATGGGTAAGAAAGTTAAGAAGGCCGTGAAGAAAGTCACCAAGTCCG...   \n",
      "44  gene-T7p23  ATGGGAACTCGCTGTTCTCTGACGGACACACGTTCTGCTACGTATG...   \n",
      "45  gene-T7p29  ATGATCGTTTCTGACATCGAAGCTAACGCCCTCTTAGAGAGCGTCA...   \n",
      "46  gene-T7p01  ATGGCTATGTCTAACATGACTTACAACAACGTTTTCGACCACGCTT...   \n",
      "47  gene-T7p12  ATGATGTACTTAATGCCATTACTCATCGTCATTGTAGGATGCCTTG...   \n",
      "48  gene-T7p24  ATGACTTACAACGTGTGGAACTTCGGGGAATCCAATGGACGCTACT...   \n",
      "49  gene-T7p44  ATGGCTAGCATGACTGGTGGACAGCAAATGGGTACTAACCAAGGTA...   \n",
      "50  gene-T7p06  ATGATGAAGCACTACGTTATGCCAATCCACACGTCCAACGGGGCAA...   \n",
      "51  gene-T7p11  ATGTTTAAGAAGGTTGGTAAATTCCTTGCGGCTTTGGCAGCTATCC...   \n",
      "52  gene-T7p35  ATGGCACTTCTTGACCTTAAACAATTCTATGAGTTACGTGAAGGCT...   \n",
      "53  gene-T7p31  ATGGCTATGACAAAGAAATTTAAAGTGTCCTTCGACGTTACCGCAA...   \n",
      "54  gene-T7p16  ATGTCAAACGTAAATACAGGTTCACTTAGTGTGGACAATAAGAAGT...   \n",
      "55  gene-T7p50  ATGAGTAAAATTGAATCTGCCCTTCAAGCGGCACAACCGGGACTCT...   \n",
      "56  gene-T7p59  ATGGCTACTCCGATAAGACCCTTGAGTTACTCGCTAAGAAGGCAAA...   \n",
      "57  gene-T7p60  ATGTTCCGCTTATTGTTGAACCTACTGCGGCATAGAGTCACCTACC...   \n",
      "58  gene-T7p04  ATGTATATGCTTACTATCGGTCTACTCACCGCTCTAGGTCTAGCTG...   \n",
      "59  gene-T7p45  ATGGCTAGCATGACTGGTGGACAGCAAATGGGTACTAACCAAGGTA...   \n",
      "\n",
      "   classification  Gene_Position_Relative  GC_Content  Seq_length  A_Content  \\\n",
      "0            late                0.688935    0.458034         417   0.266187   \n",
      "1          middle                0.561159    0.515152         924   0.250000   \n",
      "2            late                0.701655    0.505922         591   0.280880   \n",
      "3           early                0.045146    0.494048         336   0.342262   \n",
      "4            late                0.651852    0.470440        2385   0.267086   \n",
      "5            late                0.929764    0.488426         432   0.275463   \n",
      "6          middle                0.475649    0.513109         267   0.269663   \n",
      "7          middle                0.262438    0.431111         450   0.328889   \n",
      "8          middle                0.461978    0.491228         114   0.201754   \n",
      "9          middle                0.484012    0.487562         402   0.251244   \n",
      "10           late                0.887748    0.480746        1662   0.274368   \n",
      "11         middle                0.432030    0.514286         210   0.242857   \n",
      "12         middle                0.526930    0.492241        1611   0.256983   \n",
      "13         middle                0.501865    0.447837         393   0.323155   \n",
      "14          early                0.112978    0.514706        2652   0.251131   \n",
      "15         middle                0.336956    0.497653         213   0.286385   \n",
      "16         middle                0.469014    0.443137         255   0.282353   \n",
      "17           late                0.918622    0.451852         270   0.274074   \n",
      "18         middle                0.417157    0.425770         357   0.327731   \n",
      "19           late                0.955104    0.527132         258   0.236434   \n",
      "20         middle                0.425670    0.443333         300   0.303333   \n",
      "21         middle                0.175727    0.469444        1080   0.282407   \n",
      "22         middle                0.353807    0.455882         408   0.299020   \n",
      "23         middle                0.329419    0.477876         339   0.262537   \n",
      "24         middle                0.151989    0.480620         129   0.317829   \n",
      "25           late                0.201192    0.482759         261   0.249042   \n",
      "26         middle                0.220873    0.469388         147   0.299320   \n",
      "27           late                0.211834    0.441624         591   0.328257   \n",
      "28           late                0.912562    0.441176         204   0.264706   \n",
      "29           late                0.930390    0.500000         252   0.277778   \n",
      "30         middle                0.238025    0.503577         699   0.277539   \n",
      "31           late                0.252047    0.490476         420   0.280952   \n",
      "32         middle                0.154418    0.422481         258   0.306202   \n",
      "33          early                0.033928    0.455128         156   0.294872   \n",
      "34         middle                0.285625    0.456284         366   0.330601   \n",
      "35         middle                0.614017    0.483926         591   0.253807   \n",
      "36         middle                0.436613    0.452830         159   0.308176   \n",
      "37         middle                0.343491    0.485185         270   0.285185   \n",
      "38           late                0.957733    0.498580        1761   0.239637   \n",
      "39         middle                0.310840    0.488536        1701   0.267490   \n",
      "40           late                0.815585    0.495577        3957   0.261562   \n",
      "41         middle                0.273756    0.521930         456   0.234649   \n",
      "42          early                0.057666    0.493519        1080   0.281481   \n",
      "43         middle                0.492876    0.513333         300   0.286667   \n",
      "44         middle                0.292836    0.487805         123   0.317073   \n",
      "45         middle                0.385833    0.505910        2115   0.249173   \n",
      "46          early                0.029521    0.468927         354   0.296610   \n",
      "47           late                0.196184    0.444444          90   0.244444   \n",
      "48         middle                0.313218    0.490741        1512   0.265212   \n",
      "49           late                0.590029    0.538012        1197   0.227235   \n",
      "50          early                0.042968    0.481481         162   0.327160   \n",
      "51           late                0.192428    0.455128         156   0.205128   \n",
      "52         middle                0.449558    0.495017         903   0.245847   \n",
      "53         middle                0.428275    0.474510         510   0.276471   \n",
      "54         middle                0.225205    0.487179         195   0.246154   \n",
      "55           late                0.737311    0.500000        2244   0.282086   \n",
      "56           late                0.967499    0.482759         174   0.264368   \n",
      "57           late                0.988131    0.473333         150   0.173333   \n",
      "58          early                0.039237    0.458333         144   0.236111   \n",
      "59           late                0.588051    0.530829        1038   0.222543   \n",
      "\n",
      "    T_Content  G_Content  C_Content  ...  kmer_CTG  kmer_CTC  kmer_CGA  \\\n",
      "0    0.275779   0.256595   0.201439  ...  0.012048  0.014458  0.021687   \n",
      "1    0.234848   0.291126   0.224026  ...  0.027115  0.013015  0.021692   \n",
      "2    0.213198   0.265651   0.240271  ...  0.015280  0.013582  0.015280   \n",
      "3    0.163690   0.255952   0.238095  ...  0.005988  0.008982  0.020958   \n",
      "4    0.262474   0.237736   0.232704  ...  0.018464  0.012170  0.014268   \n",
      "5    0.236111   0.289352   0.199074  ...  0.018605  0.011628  0.020930   \n",
      "6    0.217228   0.280899   0.232210  ...  0.011321  0.022642  0.026415   \n",
      "7    0.240000   0.251111   0.180000  ...  0.015625  0.011161  0.013393   \n",
      "8    0.307018   0.298246   0.192982  ...  0.008929  0.017857  0.008929   \n",
      "9    0.261194   0.281095   0.206468  ...  0.017500  0.012500  0.017500   \n",
      "10   0.244886   0.255114   0.225632  ...  0.016867  0.013855  0.015060   \n",
      "11   0.242857   0.261905   0.252381  ...  0.028846  0.014423  0.009615   \n",
      "12   0.250776   0.253259   0.238982  ...  0.025482  0.022996  0.014916   \n",
      "13   0.229008   0.241730   0.206107  ...  0.007673  0.015345  0.007673   \n",
      "14   0.234163   0.264706   0.250000  ...  0.031698  0.015472  0.014340   \n",
      "15   0.215962   0.244131   0.253521  ...  0.037915  0.018957  0.023697   \n",
      "16   0.274510   0.231373   0.211765  ...  0.007905  0.011858  0.015810   \n",
      "17   0.274074   0.248148   0.203704  ...  0.011194  0.018657  0.003731   \n",
      "18   0.246499   0.260504   0.165266  ...  0.014085  0.002817  0.011268   \n",
      "19   0.236434   0.255814   0.271318  ...  0.035156  0.019531  0.011719   \n",
      "20   0.253333   0.250000   0.193333  ...  0.016779  0.010067  0.010067   \n",
      "21   0.248148   0.256481   0.212963  ...  0.029685  0.012059  0.014842   \n",
      "22   0.245098   0.254902   0.200980  ...  0.022167  0.009852  0.007389   \n",
      "23   0.259587   0.230088   0.247788  ...  0.029674  0.011869  0.011869   \n",
      "24   0.201550   0.255814   0.224806  ...  0.007874  0.023622  0.015748   \n",
      "25   0.268199   0.233716   0.249042  ...  0.011583  0.007722  0.023166   \n",
      "26   0.231293   0.238095   0.231293  ...  0.020690  0.006897  0.034483   \n",
      "27   0.230118   0.226734   0.214890  ...  0.022071  0.010187  0.015280   \n",
      "28   0.294118   0.274510   0.166667  ...  0.019802  0.009901  0.019802   \n",
      "29   0.222222   0.293651   0.206349  ...  0.024000  0.012000  0.036000   \n",
      "30   0.218884   0.280401   0.223176  ...  0.032999  0.010043  0.027260   \n",
      "31   0.228571   0.266667   0.223810  ...  0.016746  0.019139  0.014354   \n",
      "32   0.271318   0.248062   0.174419  ...  0.027344  0.000000  0.007812   \n",
      "33   0.250000   0.217949   0.237179  ...  0.045455  0.019481  0.012987   \n",
      "34   0.213115   0.256831   0.199454  ...  0.010989  0.021978  0.010989   \n",
      "35   0.262267   0.252115   0.231810  ...  0.025467  0.018676  0.018676   \n",
      "36   0.238994   0.251572   0.201258  ...  0.012739  0.019108  0.031847   \n",
      "37   0.229630   0.233333   0.251852  ...  0.026119  0.014925  0.014925   \n",
      "38   0.261783   0.260080   0.238501  ...  0.018761  0.017055  0.008528   \n",
      "39   0.243974   0.268665   0.219871  ...  0.025898  0.011183  0.016480   \n",
      "40   0.242861   0.264342   0.231236  ...  0.025032  0.018458  0.011631   \n",
      "41   0.243421   0.276316   0.245614  ...  0.024229  0.015419  0.022026   \n",
      "42   0.225000   0.264815   0.228704  ...  0.019481  0.009276  0.017625   \n",
      "43   0.200000   0.320000   0.193333  ...  0.030201  0.010067  0.003356   \n",
      "44   0.195122   0.276423   0.211382  ...  0.041322  0.024793  0.008264   \n",
      "45   0.244917   0.272813   0.233097  ...  0.023663  0.017037  0.017037   \n",
      "46   0.234463   0.245763   0.223164  ...  0.028409  0.008523  0.022727   \n",
      "47   0.311111   0.222222   0.222222  ...  0.011364  0.022727  0.011364   \n",
      "48   0.244048   0.269841   0.220899  ...  0.025166  0.010596  0.016556   \n",
      "49   0.234754   0.286550   0.251462  ...  0.041004  0.020921  0.011715   \n",
      "50   0.191358   0.222222   0.259259  ...  0.012500  0.006250  0.012500   \n",
      "51   0.339744   0.282051   0.173077  ...  0.012987  0.006494  0.000000   \n",
      "52   0.259136   0.269103   0.225914  ...  0.026637  0.012209  0.018868   \n",
      "53   0.249020   0.254902   0.219608  ...  0.021654  0.011811  0.009843   \n",
      "54   0.266667   0.266667   0.220513  ...  0.015544  0.015544  0.020725   \n",
      "55   0.217914   0.266043   0.233957  ...  0.024978  0.017395  0.018287   \n",
      "56   0.252874   0.270115   0.212644  ...  0.005814  0.011628  0.017442   \n",
      "57   0.353333   0.226667   0.246667  ...  0.033784  0.027027  0.013514   \n",
      "58   0.305556   0.236111   0.222222  ...  0.014085  0.014085  0.000000   \n",
      "59   0.246628   0.285164   0.245665  ...  0.040541  0.023166  0.010618   \n",
      "\n",
      "    kmer_CGT  kmer_CGG  kmer_CGC  kmer_CCA  kmer_CCT  kmer_CCG  kmer_CCC  \n",
      "0   0.014458  0.012048  0.002410  0.012048  0.012048  0.009639  0.007229  \n",
      "1   0.029284  0.006508  0.010846  0.010846  0.007592  0.016269  0.006508  \n",
      "2   0.010187  0.008489  0.022071  0.013582  0.008489  0.018676  0.001698  \n",
      "3   0.014970  0.011976  0.029940  0.014970  0.008982  0.017964  0.002994  \n",
      "4   0.015946  0.010911  0.008393  0.014268  0.018884  0.010491  0.006714  \n",
      "5   0.020930  0.009302  0.016279  0.002326  0.009302  0.011628  0.004651  \n",
      "6   0.022642  0.011321  0.015094  0.018868  0.007547  0.033962  0.003774  \n",
      "7   0.011161  0.006696  0.006696  0.008929  0.008929  0.008929  0.011161  \n",
      "8   0.035714  0.008929  0.000000  0.008929  0.017857  0.008929  0.000000  \n",
      "9   0.025000  0.007500  0.007500  0.005000  0.015000  0.012500  0.000000  \n",
      "10  0.017470  0.010241  0.011446  0.015663  0.008434  0.015060  0.003614  \n",
      "11  0.019231  0.019231  0.014423  0.009615  0.019231  0.014423  0.009615  \n",
      "12  0.016781  0.008701  0.010566  0.011809  0.016159  0.013673  0.008701  \n",
      "13  0.007673  0.002558  0.012788  0.017903  0.010230  0.007673  0.002558  \n",
      "14  0.015472  0.012830  0.020000  0.012075  0.013208  0.014340  0.004906  \n",
      "15  0.018957  0.000000  0.014218  0.033175  0.000000  0.009479  0.004739  \n",
      "16  0.019763  0.000000  0.007905  0.011858  0.023715  0.003953  0.007905  \n",
      "17  0.007463  0.022388  0.011194  0.011194  0.011194  0.011194  0.000000  \n",
      "18  0.000000  0.011268  0.000000  0.014085  0.005634  0.002817  0.002817  \n",
      "19  0.015625  0.003906  0.023438  0.015625  0.027344  0.011719  0.019531  \n",
      "20  0.020134  0.010067  0.003356  0.010067  0.010067  0.003356  0.006711  \n",
      "21  0.017625  0.008349  0.009276  0.009276  0.011132  0.009276  0.009276  \n",
      "22  0.009852  0.004926  0.012315  0.017241  0.002463  0.007389  0.004926  \n",
      "23  0.014837  0.002967  0.017804  0.014837  0.014837  0.008902  0.011869  \n",
      "24  0.047244  0.000000  0.015748  0.015748  0.000000  0.015748  0.000000  \n",
      "25  0.038610  0.007722  0.011583  0.003861  0.007722  0.034749  0.011583  \n",
      "26  0.006897  0.000000  0.000000  0.006897  0.013793  0.013793  0.013793  \n",
      "27  0.011885  0.003396  0.006791  0.015280  0.016978  0.011885  0.011885  \n",
      "28  0.004950  0.004950  0.004950  0.014851  0.009901  0.000000  0.000000  \n",
      "29  0.016000  0.012000  0.016000  0.000000  0.012000  0.020000  0.000000  \n",
      "30  0.017217  0.011478  0.012912  0.014347  0.011478  0.014347  0.004304  \n",
      "31  0.007177  0.009569  0.014354  0.016746  0.019139  0.007177  0.004785  \n",
      "32  0.011719  0.007812  0.000000  0.019531  0.011719  0.000000  0.000000  \n",
      "33  0.006494  0.006494  0.019481  0.006494  0.000000  0.006494  0.000000  \n",
      "34  0.010989  0.002747  0.008242  0.019231  0.008242  0.008242  0.002747  \n",
      "35  0.018676  0.010187  0.016978  0.006791  0.008489  0.020374  0.003396  \n",
      "36  0.019108  0.006369  0.006369  0.006369  0.012739  0.000000  0.000000  \n",
      "37  0.033582  0.003731  0.014925  0.011194  0.007463  0.018657  0.000000  \n",
      "38  0.025014  0.009096  0.012507  0.013076  0.019898  0.013076  0.010233  \n",
      "39  0.017657  0.008829  0.011183  0.010594  0.011772  0.013537  0.003531  \n",
      "40  0.016182  0.007838  0.011125  0.014412  0.015171  0.010619  0.008091  \n",
      "41  0.015419  0.006608  0.022026  0.017621  0.008811  0.006608  0.008811  \n",
      "42  0.012987  0.009276  0.027829  0.005566  0.006494  0.011132  0.005566  \n",
      "43  0.023490  0.020134  0.006711  0.010067  0.006711  0.020134  0.003356  \n",
      "44  0.016529  0.016529  0.008264  0.008264  0.008264  0.000000  0.008264  \n",
      "45  0.018457  0.008519  0.014671  0.009938  0.015618  0.014671  0.009938  \n",
      "46  0.017045  0.000000  0.017045  0.005682  0.014205  0.014205  0.000000  \n",
      "47  0.011364  0.000000  0.022727  0.034091  0.011364  0.000000  0.000000  \n",
      "48  0.018543  0.008609  0.011258  0.010596  0.011921  0.014570  0.003311  \n",
      "49  0.015900  0.013389  0.023431  0.015063  0.010042  0.018410  0.006695  \n",
      "50  0.025000  0.012500  0.031250  0.018750  0.012500  0.025000  0.000000  \n",
      "51  0.012987  0.019481  0.012987  0.000000  0.019481  0.000000  0.006494  \n",
      "52  0.015538  0.004440  0.007769  0.013319  0.019978  0.019978  0.006659  \n",
      "53  0.019685  0.013780  0.007874  0.009843  0.015748  0.007874  0.007874  \n",
      "54  0.025907  0.010363  0.005181  0.005181  0.020725  0.020725  0.005181  \n",
      "55  0.013381  0.010259  0.014273  0.016057  0.010259  0.011151  0.008029  \n",
      "56  0.011628  0.023256  0.005814  0.011628  0.017442  0.011628  0.011628  \n",
      "57  0.006757  0.006757  0.013514  0.000000  0.027027  0.013514  0.006757  \n",
      "58  0.000000  0.007042  0.021127  0.000000  0.014085  0.014085  0.000000  \n",
      "59  0.018340  0.014479  0.020270  0.015444  0.009653  0.016409  0.004826  \n",
      "\n",
      "[60 rows x 80 columns]\n",
      "Features table saved as ../data/dna_feature_table\n",
      "                       Geneid  \\\n",
      "0           gene-PPLUZ19_gp23   \n",
      "1           gene-PPLUZ19_gp36   \n",
      "2           gene-PPLUZ19_gp16   \n",
      "3           gene-PPLUZ19_gp40   \n",
      "4           gene-PPLUZ19_gp44   \n",
      "5           gene-PPLUZ19_gp48   \n",
      "6            gene-PPLUZ19_gp5   \n",
      "7            gene-PPLUZ19_gp6   \n",
      "8           gene-PPLUZ19_gp46   \n",
      "9         gene-PPLUZ19_gp46.1   \n",
      "10          gene-PPLUZ19_gp31   \n",
      "11           gene-PPLUZ19_gp9   \n",
      "12          gene-PPLUZ19_gp19   \n",
      "13          gene-PPLUZ19_gp24   \n",
      "14          gene-PPLUZ19_gp11   \n",
      "15          gene-PPLUZ19_gp27   \n",
      "16         gene-PPLUZ19_gp0.1   \n",
      "17           gene-PPLUZ19_gp4   \n",
      "18          gene-PPLUZ19_gp42   \n",
      "19           gene-PPLUZ19_gp7   \n",
      "20          gene-PPLUZ19_gp26   \n",
      "21          gene-PPLUZ19_gp34   \n",
      "22          gene-PPLUZ19_gp20   \n",
      "23           gene-PPLUZ19_gp8   \n",
      "24          gene-PPLUZ19_gp14   \n",
      "25          gene-PPLUZ19_gp28   \n",
      "26          gene-PPLUZ19_gp35   \n",
      "27          gene-PPLUZ19_gp41   \n",
      "28          gene-PPLUZ19_gp32   \n",
      "29          gene-PPLUZ19_gp45   \n",
      "30          gene-PPLUZ19_gp10   \n",
      "31          gene-PPLUZ19_gp13   \n",
      "32          gene-PPLUZ19_gp49   \n",
      "33          gene-PPLUZ19_gp21   \n",
      "34        gene-PPLUZ19_gp12.1   \n",
      "35          gene-PPLUZ19_gp43   \n",
      "36  gene-PPLUZ19_PPLUZ19_gp20   \n",
      "37          gene-PPLUZ19_gp12   \n",
      "38          gene-PPLUZ19_gp25   \n",
      "39           gene-PPLUZ19_gp1   \n",
      "40          gene-PPLUZ19_gp47   \n",
      "41        gene-PPLUZ19_gp25.1   \n",
      "42          gene-PPLUZ19_gp22   \n",
      "43          gene-PPLUZ19_gp37   \n",
      "44           gene-PPLUZ19_gp3   \n",
      "45          gene-PPLUZ19_gp30   \n",
      "46          gene-PPLUZ19_gp33   \n",
      "47        gene-PPLUZ19_gp13.1   \n",
      "48          gene-PPLUZ19_gp15   \n",
      "49          gene-PPLUZ19_gp39   \n",
      "50           gene-PPLUZ19_gp2   \n",
      "51          gene-PPLUZ19_gp18   \n",
      "52          gene-PPLUZ19_gp38   \n",
      "53          gene-PPLUZ19_gp29   \n",
      "\n",
      "                                          DNASequence classification  \\\n",
      "0   ATGACGAGTGAGCCGAAGGTCTACCAGATACCGCGCAGTCAACAGC...           late   \n",
      "1   ATGGCGGAATCGCAACGTGCTTCCCAAGAGCTTGGGATCAACGTCG...           late   \n",
      "2   ATGCAAGCTAAGCATAGCCGGGTGCTCGAAGGCACTAAAGAAATTC...         middle   \n",
      "3   ATGTTTAAGACCGAAGTAAAGGGACGTTACACCCTGATTCGCCGCA...           late   \n",
      "4   ATGATGCTCGATACCGCCACCGAGGCGGGCAAAGGCACCCTCGCCG...           late   \n",
      "5   ATGGCAACCTTCGCCGCTGCAACTCAGAAAGACCTCCGCGCCTTCG...           late   \n",
      "6   ATGTCCTCGCGTGATCCCTACCGCATCGGCCACCGCGTGGGGCTGG...         middle   \n",
      "7   ATGACCCTCGTGGCCACCGTAGTAGACAGCGCGCACAACCTGGAAG...         middle   \n",
      "8   ATGCCTCGAACCATCGTCGCAATACTCGTCCTTGCTGTGGTAGCCC...           late   \n",
      "9   GTGCCCTGGACCAGAACCGCCCGTGGGCTGACCGGCCTGTGCCTGC...           late   \n",
      "10  ATGACCCAACAGAACGAACAGCAACTGCCGCCGGGCCTCGCTAACC...           late   \n",
      "11  ATGATCCGTACCCATACCCACAACACCGACCGCGAACAGCGCCGCA...         middle   \n",
      "12  ATGACGACTATCCGAATCCTCGACCTCGAAACCGAGAGCTACGAGC...         middle   \n",
      "13  ATGAGCAAACTCCGCAAGCAATTCACCAATGAATACCTGCGCAACG...           late   \n",
      "14  ATGAGCAAGCGCAACCCCGAACACATCAACGGCACCGTTCGTAGCG...         middle   \n",
      "15  ATGGCTACTATGAAGACCCACCGCCCTACGGTTATGTCACCCACAG...           late   \n",
      "16  ATGAGTGCTGACGATCCCGAAGAGTATGCGCAGGACAAGCCAGAAC...         middle   \n",
      "17  ATGCAAGCTTTGAATACCCTGTTGATTGCAATCCCCAAGGACCCGA...         middle   \n",
      "18  ATGAGTAAAAAGCAGACCGCGAGTGCTGAGCGGCTGGGCCTGCTTC...           late   \n",
      "19  ATGGTCACCCGTACTGTATACGTCACGCCTGAAGACCCGACGCCGC...         middle   \n",
      "20  ATGGACCTGATACAGCAGCAGATCGCCCACGAAGAGGCCCTGGTCG...           late   \n",
      "21  ATGAGCTACAAGCAATCCGCGTATCCCAATCTGCTGATGGGTGTGA...           late   \n",
      "22  GTGAAAGAACTGCACCCGCTGCACACGCCTGAGTTCGTCAAGACAT...           late   \n",
      "23  ATGGCCGTGGCAATACTCATCCTGGCCGTGTGGTTGATCGGCGGCG...         middle   \n",
      "24  ATGGCACTGCGCCGTGACTCCTGGCTAAAGCAAGCGCAATCCCTGG...         middle   \n",
      "25  ATGAACAAGTCCATCTGGCGAGTCCACGCAAAGGCCGGCACTCCCT...           late   \n",
      "26  ATGGCTTTCTGGCTACCACTATTGGCCGCTGGCGGCATGTCCGCCC...           late   \n",
      "27  ATGGCACTGATCTACGACTTCAACCCAGACCTTGATCCGAAGGCTA...           late   \n",
      "28  ATGAGCTTTCTGAACGACCTGACCCGTCCGAACTACGCTGGCAAGA...           late   \n",
      "29  GTGAACAAGCCCCTGCGCGGCGCAGCCCTTGCGGCTGCCCTCGCCG...           late   \n",
      "30  ATGACCAACGTCAACACCACCACCGAGACCACCGCCGCTGCTGTCC...         middle   \n",
      "31  GTGCTGGCCCTCGGTGCCTTCGACCTGTCCGGCCTGATGGTAGGTT...         middle   \n",
      "32  ATGAGCAAAGCCAAACTACGAGTCATCGCCGACACCCCGGAGCTGG...           late   \n",
      "33  ATGACTCAACAACTCAACGCTCTGCAAGCCGCCCTCGCCCTGGCCA...           late   \n",
      "34  ATGCAATACCACTTCACGCATTACAACGGATACCGCTTCGGCGTCG...         middle   \n",
      "35  ATGACGCCGCAAGAACGATTCCAGATAGCCCACGAGGTGCGGGACA...           late   \n",
      "36  ATGAGCAAGCGCGACGTGGTACTGGATATCGAGAAAGGCATCTGGC...         middle   \n",
      "37  ATGCAGTGCAAAGACCTTTACACGAACCTCGCGTCGGGCATGTTCA...         middle   \n",
      "38  ATGAGCCTAGCATTCCCGGACTCTTACGAGTCGACGATCACGACTG...           late   \n",
      "39  ATGGCACACTTCAAGGCTAAGGCTCCCAAGTCGCCCTTTGCTGCTC...          early   \n",
      "40  ATGGCGAACACCCGCGAGCAGTACCTCGCTGGCCGTAACACCGGCC...           late   \n",
      "41  ATGAGTAAAATCTGTTGGTGTACCCGACCGCACGAGACCGATGAAG...           late   \n",
      "42  ATGCGCCTGCCATCGGAAGAGTTCCTGGCAGGACTATCCGCGCAGT...           late   \n",
      "43  ATGGCAAAGCAATTCAAGGGCCGCATGACGCCCAAGTATCCCCTTG...           late   \n",
      "44  ATGACCAACGCAATCTCCAAAACCGTAATCGCATTCCGTGGCACCG...          early   \n",
      "45  ATGAAAACCACCGCAGCTATGCTGTGGGAGAAACTTCGGGACGGGA...           late   \n",
      "46  ATGCTACTACTCGACGCAGTGAATGTCATCCTGCGCAAGATCGGCG...           late   \n",
      "47  ATGACGTATGAGGTGATGACATGGCTTATCGAGAACAAACCGCTGG...         middle   \n",
      "48  ATGATGGGGCCGGAGACGTGCTTTGTCATCGACTGGATCGAGCAAT...         middle   \n",
      "49  ATGCGCGGCATTATCGCGGGCATCATGGCCTCCCAAATTCGACGGC...           late   \n",
      "50  ATGCCACGTGTGAATGAACTGACACCGCGTCAACGCAAGGCCGCCA...          early   \n",
      "51  ATGAGAATGCCAACCGAAGAAGAACGCATGATCCGCTGTTTACTGG...         middle   \n",
      "52  GTGGCTCGGTTCAAGAATCCCGAGACCATCCACGTTGCAGATGGGG...           late   \n",
      "53  GTGTTGGGCAAGACCATCATCGGCAAACTCGCTGATGGCCTGCTGG...           late   \n",
      "\n",
      "    Gene_Position_Relative  GC_Content  Seq_length  A_Content  T_Content  \\\n",
      "0                 0.382704    0.607710         441   0.251701   0.140590   \n",
      "1                 0.708414    0.626622        2697   0.223582   0.149796   \n",
      "2                 0.232135    0.619968         621   0.188406   0.191626   \n",
      "3                 0.869753    0.593819         906   0.218543   0.187638   \n",
      "4                 0.945026    0.636816         201   0.179104   0.184080   \n",
      "5                 0.978116    0.617886         246   0.223577   0.158537   \n",
      "6                 0.083861    0.666667         225   0.186667   0.146667   \n",
      "7                 0.093529    0.590038         261   0.226054   0.183908   \n",
      "8                 0.961146    0.657576         330   0.166667   0.175758   \n",
      "9                 0.964637    0.651741         201   0.169154   0.179104   \n",
      "10                0.557523    0.658411         969   0.188854   0.152735   \n",
      "11                0.111877    0.625000         288   0.208333   0.166667   \n",
      "12                0.296041    0.617162        2424   0.220710   0.162129   \n",
      "13                0.399697    0.603629        1047   0.228271   0.168099   \n",
      "14                0.132245    0.644444         360   0.211111   0.144444   \n",
      "15                0.491595    0.583333         252   0.246032   0.170635   \n",
      "16                0.038555    0.542636         129   0.279070   0.178295   \n",
      "17                0.077363    0.626016         369   0.192412   0.181572   \n",
      "18                0.897630    0.584967         306   0.248366   0.166667   \n",
      "19                0.099844    0.632302         291   0.202749   0.164948   \n",
      "20                0.456646    0.618464        2448   0.208333   0.173203   \n",
      "21                0.636470    0.621121        2481   0.205159   0.173720   \n",
      "22                0.327363    0.628205         312   0.208333   0.163462   \n",
      "23                0.105814    0.610390         231   0.164502   0.225108   \n",
      "24                0.187379    0.602424         825   0.200000   0.197576   \n",
      "25                0.499908    0.592827         474   0.225738   0.181435   \n",
      "26                0.671213    0.642857         546   0.206960   0.150183   \n",
      "27                0.887182    0.589109         606   0.224422   0.186469   \n",
      "28                0.581427    0.626984        1008   0.209325   0.163690   \n",
      "29                0.952788    0.616977         483   0.219462   0.163561   \n",
      "30                0.121751    0.654676         417   0.211031   0.134293   \n",
      "31                0.170364    0.642361         288   0.159722   0.197917   \n",
      "32                0.983627    0.597222         216   0.226852   0.175926   \n",
      "33                0.344241    0.641905        1050   0.211429   0.146667   \n",
      "34                0.161661    0.583333         420   0.228571   0.188095   \n",
      "35                0.922086    0.620709        1806   0.207087   0.172204   \n",
      "36                0.250138    0.640295         948   0.193038   0.166667   \n",
      "37                0.145724    0.639506         810   0.192593   0.167901   \n",
      "38                0.416207    0.596774         372   0.239247   0.163978   \n",
      "39                0.043584    0.575439         285   0.273684   0.150877   \n",
      "40                0.970584    0.666667         315   0.203175   0.130159   \n",
      "41                0.424313    0.615385         351   0.205128   0.179487   \n",
      "42                0.367089    0.621019         942   0.219745   0.159236   \n",
      "43                0.785547    0.649975        4014   0.189836   0.160189   \n",
      "44                0.058510    0.588889         540   0.248148   0.162963   \n",
      "45                0.528727    0.626223        1533   0.202218   0.171559   \n",
      "46                0.601566    0.623423         555   0.212613   0.163964   \n",
      "47                0.174887    0.604651         129   0.209302   0.186047   \n",
      "48                0.211560    0.618090        1194   0.208543   0.173367   \n",
      "49                0.854253    0.601307         459   0.220044   0.178649   \n",
      "50                0.049463    0.627193         228   0.228070   0.144737   \n",
      "51                0.264605    0.563863         321   0.246106   0.190031   \n",
      "52                0.840337    0.601852         756   0.211640   0.186508   \n",
      "53                0.507716    0.626812         276   0.246377   0.126812   \n",
      "\n",
      "    G_Content  C_Content  ...  kmer_CTG  kmer_CTC  kmer_CGA  kmer_CGT  \\\n",
      "0    0.306122   0.301587  ...  0.027335  0.006834  0.041002  0.011390   \n",
      "1    0.325176   0.301446  ...  0.021150  0.012616  0.027829  0.014100   \n",
      "2    0.339775   0.280193  ...  0.027464  0.016155  0.019386  0.021002   \n",
      "3    0.291391   0.302428  ...  0.025442  0.015487  0.018805  0.016593   \n",
      "4    0.293532   0.343284  ...  0.015075  0.045226  0.020101  0.020101   \n",
      "5    0.260163   0.357724  ...  0.024590  0.024590  0.032787  0.012295   \n",
      "6    0.320000   0.346667  ...  0.026906  0.004484  0.040359  0.022422   \n",
      "7    0.283525   0.306513  ...  0.023166  0.015444  0.011583  0.023166   \n",
      "8    0.318182   0.339394  ...  0.042683  0.009146  0.006098  0.030488   \n",
      "9    0.288557   0.363184  ...  0.075377  0.000000  0.005025  0.010050   \n",
      "10   0.305470   0.352941  ...  0.026887  0.014478  0.019648  0.016546   \n",
      "11   0.305556   0.319444  ...  0.031469  0.006993  0.027972  0.020979   \n",
      "12   0.313944   0.303218  ...  0.024360  0.012386  0.032618  0.015277   \n",
      "13   0.301815   0.301815  ...  0.033493  0.007656  0.025837  0.014354   \n",
      "14   0.313889   0.330556  ...  0.044693  0.005587  0.022346  0.025140   \n",
      "15   0.293651   0.289683  ...  0.016000  0.008000  0.032000  0.012000   \n",
      "16   0.279070   0.263566  ...  0.015748  0.000000  0.023622  0.000000   \n",
      "17   0.308943   0.317073  ...  0.029973  0.010899  0.024523  0.024523   \n",
      "18   0.277778   0.307190  ...  0.029605  0.006579  0.042763  0.009868   \n",
      "19   0.319588   0.312715  ...  0.031142  0.017301  0.017301  0.024221   \n",
      "20   0.308824   0.309641  ...  0.027800  0.014718  0.028618  0.012265   \n",
      "21   0.301894   0.319226  ...  0.028237  0.012908  0.026220  0.017346   \n",
      "22   0.298077   0.330128  ...  0.038710  0.006452  0.012903  0.029032   \n",
      "23   0.259740   0.350649  ...  0.026201  0.017467  0.017467  0.026201   \n",
      "24   0.321212   0.281212  ...  0.037667  0.003645  0.018226  0.007290   \n",
      "25   0.299578   0.293249  ...  0.025424  0.016949  0.029661  0.016949   \n",
      "26   0.336996   0.305861  ...  0.025735  0.016544  0.018382  0.007353   \n",
      "27   0.305281   0.283828  ...  0.014901  0.014901  0.028146  0.019868   \n",
      "28   0.296627   0.330357  ...  0.035785  0.011928  0.031809  0.022863   \n",
      "29   0.279503   0.337474  ...  0.033264  0.012474  0.014553  0.020790   \n",
      "30   0.294964   0.359712  ...  0.031325  0.004819  0.031325  0.024096   \n",
      "31   0.357639   0.284722  ...  0.059441  0.006993  0.020979  0.017483   \n",
      "32   0.287037   0.310185  ...  0.028037  0.009346  0.032710  0.009346   \n",
      "33   0.292381   0.349524  ...  0.022901  0.016221  0.026718  0.009542   \n",
      "34   0.309524   0.273810  ...  0.031100  0.009569  0.028708  0.007177   \n",
      "35   0.300111   0.320598  ...  0.030488  0.014967  0.031042  0.017738   \n",
      "36   0.320675   0.319620  ...  0.025370  0.012685  0.025370  0.023256   \n",
      "37   0.343210   0.296296  ...  0.030941  0.011139  0.017327  0.024752   \n",
      "38   0.330645   0.266129  ...  0.035135  0.005405  0.037838  0.002703   \n",
      "39   0.315789   0.259649  ...  0.017668  0.010601  0.010601  0.024735   \n",
      "40   0.304762   0.361905  ...  0.022364  0.015974  0.019169  0.019169   \n",
      "41   0.373219   0.242165  ...  0.034384  0.005731  0.037249  0.011461   \n",
      "42   0.326964   0.294055  ...  0.022340  0.009574  0.032979  0.013830   \n",
      "43   0.330095   0.319880  ...  0.027418  0.014955  0.022183  0.015454   \n",
      "44   0.285185   0.303704  ...  0.033457  0.007435  0.026022  0.018587   \n",
      "45   0.313764   0.312459  ...  0.031352  0.015023  0.028086  0.016329   \n",
      "46   0.299099   0.324324  ...  0.025316  0.010850  0.021700  0.009042   \n",
      "47   0.310078   0.294574  ...  0.031496  0.007874  0.007874  0.015748   \n",
      "48   0.334171   0.283920  ...  0.029362  0.010067  0.026007  0.012584   \n",
      "49   0.281046   0.320261  ...  0.030635  0.021882  0.008753  0.010941   \n",
      "50   0.271930   0.355263  ...  0.039823  0.008850  0.022124  0.013274   \n",
      "51   0.280374   0.283489  ...  0.043887  0.003135  0.021944  0.009404   \n",
      "52   0.291005   0.310847  ...  0.019894  0.021220  0.034483  0.029178   \n",
      "53   0.311594   0.315217  ...  0.021898  0.014599  0.025547  0.007299   \n",
      "\n",
      "    kmer_CGG  kmer_CGC  kmer_CCA  kmer_CCT  kmer_CCG  kmer_CCC  \n",
      "0   0.018223  0.025057  0.013667  0.020501  0.031891  0.015945  \n",
      "1   0.023748  0.028200  0.019295  0.018182  0.029685  0.011503  \n",
      "2   0.029079  0.022617  0.011309  0.017771  0.029079  0.016155  \n",
      "3   0.027655  0.025442  0.013274  0.017699  0.034292  0.024336  \n",
      "4   0.025126  0.050251  0.015075  0.020101  0.040201  0.025126  \n",
      "5   0.012295  0.049180  0.016393  0.036885  0.053279  0.012295  \n",
      "6   0.026906  0.040359  0.022422  0.022422  0.044843  0.022422  \n",
      "7   0.019305  0.030888  0.019305  0.023166  0.038610  0.015444  \n",
      "8   0.021341  0.027439  0.033537  0.033537  0.027439  0.033537  \n",
      "9   0.015075  0.040201  0.020101  0.050251  0.035176  0.045226  \n",
      "10  0.026887  0.039297  0.023785  0.018614  0.053775  0.022751  \n",
      "11  0.013986  0.024476  0.027972  0.013986  0.038462  0.027972  \n",
      "12  0.026012  0.023534  0.020231  0.016102  0.032205  0.016928  \n",
      "13  0.017225  0.030622  0.019139  0.020096  0.024880  0.015311  \n",
      "14  0.008380  0.041899  0.027933  0.016760  0.033520  0.019553  \n",
      "15  0.016000  0.020000  0.024000  0.024000  0.020000  0.024000  \n",
      "16  0.000000  0.015748  0.031496  0.023622  0.023622  0.023622  \n",
      "17  0.019074  0.043597  0.016349  0.019074  0.038147  0.024523  \n",
      "18  0.013158  0.032895  0.016447  0.013158  0.032895  0.013158  \n",
      "19  0.020761  0.031142  0.013841  0.017301  0.041522  0.020761  \n",
      "20  0.025756  0.026165  0.022895  0.019215  0.023303  0.019215  \n",
      "21  0.030254  0.020573  0.025010  0.022993  0.027834  0.016136  \n",
      "22  0.029032  0.025806  0.025806  0.022581  0.029032  0.022581  \n",
      "23  0.013100  0.043668  0.013100  0.034934  0.052402  0.021834  \n",
      "24  0.027947  0.029162  0.017011  0.019441  0.020656  0.019441  \n",
      "25  0.023305  0.016949  0.021186  0.016949  0.025424  0.021186  \n",
      "26  0.033088  0.034926  0.018382  0.022059  0.029412  0.011029  \n",
      "27  0.019868  0.019868  0.014901  0.019868  0.024834  0.008278  \n",
      "28  0.013917  0.040755  0.024851  0.026839  0.036779  0.011928  \n",
      "29  0.016632  0.037422  0.033264  0.024948  0.018711  0.016632  \n",
      "30  0.019277  0.043373  0.038554  0.019277  0.040964  0.009639  \n",
      "31  0.024476  0.013986  0.006993  0.031469  0.031469  0.013986  \n",
      "32  0.014019  0.037383  0.009346  0.014019  0.046729  0.023364  \n",
      "33  0.037214  0.024809  0.023855  0.020038  0.039122  0.030534  \n",
      "34  0.023923  0.019139  0.014354  0.021531  0.028708  0.011962  \n",
      "35  0.023836  0.028825  0.019401  0.018847  0.030488  0.020510  \n",
      "36  0.024313  0.038055  0.017970  0.021142  0.035941  0.016913  \n",
      "37  0.034653  0.032178  0.018564  0.014851  0.027228  0.016089  \n",
      "38  0.018919  0.024324  0.010811  0.013514  0.032432  0.013514  \n",
      "39  0.010601  0.038869  0.017668  0.010601  0.007067  0.014134  \n",
      "40  0.031949  0.031949  0.028754  0.022364  0.054313  0.022364  \n",
      "41  0.034384  0.022923  0.002865  0.008596  0.031519  0.008596  \n",
      "42  0.028723  0.028723  0.017021  0.020213  0.026596  0.005319  \n",
      "43  0.031655  0.031655  0.021934  0.021685  0.034646  0.014706  \n",
      "44  0.007435  0.031599  0.031599  0.014870  0.033457  0.009294  \n",
      "45  0.015676  0.038537  0.018942  0.020901  0.032005  0.012410  \n",
      "46  0.027125  0.028933  0.027125  0.016275  0.032550  0.028933  \n",
      "47  0.031496  0.039370  0.000000  0.023622  0.023622  0.023622  \n",
      "48  0.029362  0.025168  0.015940  0.020134  0.026007  0.013423  \n",
      "49  0.037199  0.019694  0.032823  0.024070  0.017505  0.019694  \n",
      "50  0.004425  0.061947  0.017699  0.013274  0.053097  0.026549  \n",
      "51  0.009404  0.028213  0.028213  0.012539  0.031348  0.021944  \n",
      "52  0.014589  0.027851  0.019894  0.014589  0.034483  0.011936  \n",
      "53  0.029197  0.036496  0.014599  0.021898  0.032847  0.007299  \n",
      "\n",
      "[54 rows x 80 columns]\n",
      "Features table saved as ../data/dna_feature_table\n",
      "               Geneid                                        DNASequence  \\\n",
      "0   gene-VPVV882_gp17  ATGGCACTGGATCCGGAAAAATTAGCGCAAGACATCGAGGCCGCCA...   \n",
      "1   gene-VPVV882_gp16  ATGCGGACGACGAGAGCCAGTACGAGAGGGTGCCTGATGCGTGATG...   \n",
      "2   gene-VPVV882_gp71  ATGGAACCTTTGACCATTGCGCTCGGCCTGGCCAAGTTGACCGGGC...   \n",
      "3   gene-VPVV882_gp11  ATGATTGGTGACGACGACTTCGACACCTTTTACGACCCGGACGACT...   \n",
      "4   gene-VPVV882_gp69  ATGGATTTGAGCACCCTTATCGCCCTGCTGACTCTGATTGTGCTGG...   \n",
      "..                ...                                                ...   \n",
      "59  gene-VPVV882_gp49  ATGCATATCGCAAAGCAAATGATGGCCGAGTTCGGCAAGCAGCTGG...   \n",
      "60  gene-VPVV882_gp64  ATGGAGCTGGACTCGGTAAAGCAAGGCTCAGCGTGGAATTTGCACT...   \n",
      "61  gene-VPVV882_gp05  ATGAGTGAGCCGATAAAACTGCAAGCCGGTGACAGCATTACCTGGT...   \n",
      "62  gene-VPVV882_gp48  ATGGGTATTAATTGCAGAGGAGTGGCCAGCATGAATTGGAACTGGA...   \n",
      "63  gene-VPVV882_gp01  ATGGCATCCTCGGTATCAATCTCAACGGCGACCCCGCGGCCGCCTC...   \n",
      "\n",
      "   classification  Gene_Position_Relative  GC_Content  Seq_length  A_Content  \\\n",
      "0            late                0.233579    0.585859         198   0.242424   \n",
      "1            late                0.222138    0.585546         678   0.218289   \n",
      "2            late                0.986255    0.581056         549   0.209472   \n",
      "3            late                0.182815    0.578512         363   0.214876   \n",
      "4            late                0.960756    0.571429         273   0.219780   \n",
      "..            ...                     ...         ...         ...        ...   \n",
      "59           late                0.640783    0.567100         231   0.268398   \n",
      "60           late                0.924994    0.609929         282   0.195035   \n",
      "61           late                0.076917    0.593220         531   0.224105   \n",
      "62           late                0.637040    0.504065         123   0.276423   \n",
      "63           late                0.006336    0.500000         222   0.243243   \n",
      "\n",
      "    T_Content  G_Content  C_Content  ...  kmer_CTG  kmer_CTC  kmer_CGA  \\\n",
      "0    0.171717   0.323232   0.262626  ...  0.030612  0.015306  0.015306   \n",
      "1    0.196165   0.337758   0.247788  ...  0.025148  0.008876  0.028107   \n",
      "2    0.209472   0.304189   0.276867  ...  0.032907  0.012797  0.023766   \n",
      "3    0.206612   0.305785   0.272727  ...  0.024931  0.013850  0.049861   \n",
      "4    0.208791   0.274725   0.296703  ...  0.029520  0.036900  0.025830   \n",
      "..        ...        ...        ...  ...       ...       ...       ...   \n",
      "59   0.164502   0.298701   0.268398  ...  0.008734  0.004367  0.030568   \n",
      "60   0.195035   0.308511   0.301418  ...  0.017857  0.025000  0.007143   \n",
      "61   0.182674   0.305085   0.288136  ...  0.035917  0.005671  0.020794   \n",
      "62   0.219512   0.317073   0.186992  ...  0.041322  0.000000  0.000000   \n",
      "63   0.256757   0.207207   0.292793  ...  0.000000  0.031818  0.022727   \n",
      "\n",
      "    kmer_CGT  kmer_CGG  kmer_CGC  kmer_CCA  kmer_CCT  kmer_CCG  kmer_CCC  \n",
      "0   0.005102  0.015306  0.030612  0.030612  0.005102  0.025510  0.005102  \n",
      "1   0.014793  0.032544  0.023669  0.013314  0.008876  0.031065  0.004438  \n",
      "2   0.005484  0.021938  0.036563  0.012797  0.020110  0.018282  0.007313  \n",
      "3   0.016620  0.008310  0.005540  0.019391  0.016620  0.019391  0.008310  \n",
      "4   0.003690  0.018450  0.025830  0.025830  0.022140  0.018450  0.011070  \n",
      "..       ...       ...       ...       ...       ...       ...       ...  \n",
      "59  0.013100  0.021834  0.017467  0.034934  0.000000  0.026201  0.017467  \n",
      "60  0.021429  0.025000  0.039286  0.021429  0.010714  0.017857  0.003571  \n",
      "61  0.011342  0.026465  0.030246  0.015123  0.015123  0.034026  0.015123  \n",
      "62  0.016529  0.016529  0.008264  0.016529  0.016529  0.008264  0.000000  \n",
      "63  0.000000  0.022727  0.018182  0.018182  0.036364  0.013636  0.031818  \n",
      "\n",
      "[64 rows x 80 columns]\n",
      "Features table saved as ../data/dna_feature_table\n",
      "            Geneid                                        DNASequence  \\\n",
      "0   gene-M172_gp35  ATGAGTGCAAGAGACATTCCATTATTCACTCAAGAGCAATATGATT...   \n",
      "1   gene-M172_gp24  ATGTATTTGGAGAATGAGTTCAAGTTAGAAGGCGAACAGCTACGTT...   \n",
      "2   gene-M172_gp16  ATGGATTGTATTGATCACGGTATGGCTGGTGATAAAGACGGTTACG...   \n",
      "3   gene-M172_gp37  ATGAAGTCCAAAGGAAATGATTTTACAAAGACTATTCGAGCTTTGT...   \n",
      "4   gene-M172_gp34  ATGTCTGATCTATACCAACGCCAAATAGCTCTTGAAGAATCATACA...   \n",
      "5   gene-M172_gp48  ATGGGGGTGGCTGGTGCAACAGTCACCGCCCTAGTGCCTTCTGGTA...   \n",
      "6   gene-M172_gp39  ATGGCAGGTGCTAATTACTTCGCAGATGGTTCAACACGATTTCATT...   \n",
      "7   gene-M172_gp46  ATGGCTGGACTTTTTACAGGAACACAAGAATCTAAATTGTTACCAG...   \n",
      "8   gene-M172_gp13  ATGCGTGGGTATAATTATATTGCGGGTGATCGGGTTTACCGTACAG...   \n",
      "9   gene-M172_gp27  ATGAACGCATTATTAGCACAAGCAGCAGCAGCAATCGAAGCAGGTA...   \n",
      "10  gene-M172_gp23  ATGAAACATAATTGGATTATTCTGGATTTTGAGGTGGAGAACTATG...   \n",
      "11  gene-M172_gp22  ATGCTAACCGACTGTATAGATCATGGTAAAACTAAATCACTAAGTC...   \n",
      "12  gene-M172_gp26  ATGTTTAGTGAAGAACTACATACATTGTTTCAGTATGATCCAGACA...   \n",
      "13  gene-M172_gp06  ATGTCACATATAATCTATGGTATCGAACCTATTATGACTGTAATTG...   \n",
      "14  gene-M172_gp15  ATGAAGCGTAAGAATTGTGTTGCAGGTCAGTGTGTTCAAGTTAAGA...   \n",
      "15  gene-M172_gp33  ATGAGTTATAAGGTAGGTTTAATTGGACTAGCAGGTTCAGGTAAAG...   \n",
      "16  gene-M172_gp54  ATGAAGGTAAATCTATCAGCACTAAAGAAGATCAAGCATAGTGATG...   \n",
      "17  gene-M172_gp01  ATGAGACAATTAGCAGATCATATCAATAGTACGTACCTATCAAAAG...   \n",
      "18  gene-M172_gp19  ATGGCTATTATATGTGTTATTCCTAAGACGGATGCGGAGTGTAATC...   \n",
      "19  gene-M172_gp28  ATGTACACTTATATCGAAGATGAAGACTTTCAACCAGCAACACCAA...   \n",
      "20  gene-M172_gp47  ATGAATATACTACGCTCATTTACAGAGACAGTGGTGACTACACCTA...   \n",
      "21  gene-M172_gp31  ATGACAGCACGTAAGATTTCCAGAGGTCAACTAAGACCTATTGCTA...   \n",
      "22  gene-M172_gp11  ATGACTACTGAGAACAAATACCCTGACTTCTGTGTTGCTAACTATT...   \n",
      "23  gene-M172_gp45  ATGGCTACTCAAATTCAGATGCCAAATGTACGTGATGTACAAGTAA...   \n",
      "24  gene-M172_gp09  ATGGACGTAAGACTTCAACAAGTTATTGAGCAGTACAAAGATAAGA...   \n",
      "25  gene-M172_gp20  ATGTCTGACCGTGAACCAATTGATCGGAATGTGTTACATGCACTGA...   \n",
      "26  gene-M172_gp41  ATGGCTGGTGCTATTGGTCCACCGATTGTTTCAGAAGCTATTACAG...   \n",
      "27  gene-M172_gp29  ATGAATAATATTCTATCACGATTTGGCGTAACCTCAGACAGTATCT...   \n",
      "28  gene-M172_gp12  ATGTCAATTAAAGGTTATAGTGTTGCACAAGCTTATGTTGATGAGC...   \n",
      "29  gene-M172_gp36  ATGAGCTGGTTGAGTAAAGCTTTAGGTAGTGTTATGGGTATGTTCG...   \n",
      "30  gene-M172_gp10  ATGACTAAACGTAACGCTTATCCTGTTGGTAAATTTGATAATGTAG...   \n",
      "31  gene-M172_gp42  ATGACTCTACTTGAAGCTGTAAATGCTATCCTACCTTATTTAGGAC...   \n",
      "32  gene-M172_gp53  ATGACAGAACAAGAAATTAAAGAAAGACTAACTCGTAATTGGTTAG...   \n",
      "33  gene-M172_gp38  ATGAGTGAATTTAATACAGGTGGTCAAGGAAACCCTCAAGAAAATA...   \n",
      "34  gene-M172_gp14  ATGAAATTCGGTAAAGTTAAAGTAGGTATGCGTGTTCGTATTAAGA...   \n",
      "35  gene-M172_gp43  ATGATTCTTGAGGGAGTGTACCCGTCATTCTTGAAAGGTGTATCAC...   \n",
      "36  gene-M172_gp50  ATGATTCTGACTAAAGACGGGTTTAGTATTATCCGTAATGAACTAT...   \n",
      "37  gene-M172_gp17  TTGGGTCAGAAACGTAGGGTATGGCATGGAGCAGAGCAGACTAAAG...   \n",
      "38  gene-M172_gp30  ATGAAGACTTCTTGTATTGATCACGGAAGACGTGGTAATCGGGATG...   \n",
      "39  gene-M172_gp44  ATGGCTATGAACATGCAAGGTGGTATGCAAGGCGCACAGACAGGTG...   \n",
      "40  gene-M172_gp40  ATGGCAGGTGCATTATGTGAACCAATCGTATCTTTAGCATTAACTC...   \n",
      "41  gene-M172_gp07  ATGTGGACTATGAAGAATACTTCATCGCATGTTAGAGACACCTATA...   \n",
      "42  gene-M172_gp32  ATGTCTAAAGTAGTATTCAGCAGTTTAAAAGAATGGAAACAACGTG...   \n",
      "43  gene-M172_gp04  ATGGAACCTACTGTTAAACAACTGCAAGACACTATTGCACAACTTA...   \n",
      "44  gene-M172_gp21  ATGAATAGATTTGATTCAATTGAGGAATACTTAGCTAAGACTAAGA...   \n",
      "45  gene-M172_gp03  ATGAATCTATTAATCACTGTTGAAGATATTAACAAAGCAATCGCAT...   \n",
      "46  gene-M172_gp49  ATGAACATGCTAGATCAGTTCTACGCAGTTCTGATCTATGTATGGT...   \n",
      "47  gene-M172_gp52  ATGATTGATGCGACGTTCCGAGAACGCTTTAAAAGATTACGGGCAC...   \n",
      "48  gene-M172_gp25  ATGGTTAAAATTAAATATGATCCAGAGAACCCACTAACACCAGAAC...   \n",
      "49  gene-M172_gp05  ATGACGGACCGTAAATACGACTACCAATCACATGCACACATACCTA...   \n",
      "50  gene-M172_gp02  ATGTCTAATCTTAATATGCCACTAGCTGTATTACGTCCAGAATTGA...   \n",
      "51  gene-M172_gp51  ATGGCAGGTAAGAAAACAGGGGCTAGTGTTAGCCGTCTATGCTTAT...   \n",
      "52  gene-M172_gp18  ATGTTATACAACGGTATTAATATTGACGTAAACGCTGAACAACACA...   \n",
      "53  gene-M172_gp08  TTGGAGGCTTTAATGGCTAACGTACTTACAACTGGTTTATTCTGTT...   \n",
      "\n",
      "   classification  Gene_Position_Relative  GC_Content  Seq_length  A_Content  \\\n",
      "0            late                0.510158    0.343434         198   0.388889   \n",
      "1          middle                0.289534    0.383562         438   0.347032   \n",
      "2          middle                0.156264    0.393333         450   0.328889   \n",
      "3            late                0.537016    0.389852        1557   0.323699   \n",
      "4            late                0.476662    0.391232        2418   0.308933   \n",
      "5            late                0.903497    0.465517         174   0.235632   \n",
      "6            late                0.588669    0.420543        1032   0.283915   \n",
      "7            late                0.806519    0.410778        3099   0.302033   \n",
      "8          middle                0.138343    0.411458         192   0.286458   \n",
      "9          middle                0.351689    0.389450         891   0.358025   \n",
      "10         middle                0.274529    0.375145         861   0.310105   \n",
      "11         middle                0.259121    0.357942         447   0.360179   \n",
      "12         middle                0.335048    0.402083         480   0.333333   \n",
      "13          early                0.081806    0.402010         597   0.301508   \n",
      "14         middle                0.147944    0.428571         210   0.295238   \n",
      "15           late                0.440085    0.365591         651   0.324117   \n",
      "16           late                0.986962    0.382353         204   0.362745   \n",
      "17          early                0.023136    0.338164         207   0.367150   \n",
      "18         middle                0.190494    0.367089         237   0.316456   \n",
      "19         middle                0.365011    0.375000         120   0.333333   \n",
      "20           late                0.871068    0.369751        2334   0.328192   \n",
      "21         middle                0.404907    0.421769         441   0.337868   \n",
      "22         middle                0.126988    0.388128         438   0.319635   \n",
      "23           late                0.735380    0.395357        2886   0.322938   \n",
      "24          early                0.103828    0.423559         399   0.285714   \n",
      "25         middle                0.208984    0.409546        1299   0.304850   \n",
      "26           late                0.610359    0.391156         294   0.336735   \n",
      "27         middle                0.378523    0.384535         957   0.335423   \n",
      "28         middle                0.134408    0.392857         168   0.357143   \n",
      "29           late                0.515396    0.448413         252   0.321429   \n",
      "30         middle                0.116013    0.386503         489   0.312883   \n",
      "31           late                0.623800    0.374332         561   0.320856   \n",
      "32           late                0.983904    0.325926         135   0.437037   \n",
      "33           late                0.565865    0.416957         861   0.332172   \n",
      "34         middle                0.143108    0.347032         219   0.360731   \n",
      "35           late                0.657817    0.380454        2292   0.307592   \n",
      "36           late                0.920920    0.388889         558   0.306452   \n",
      "37         middle                0.171246    0.391534         756   0.313492   \n",
      "38         middle                0.394595    0.381279         438   0.326484   \n",
      "39           late                0.692924    0.421131         672   0.303571   \n",
      "40           late                0.604409    0.440860         186   0.338710   \n",
      "41          early                0.090814    0.412698         189   0.296296   \n",
      "42           late                0.421311    0.387821         936   0.308761   \n",
      "43          early                0.060685    0.402667         375   0.298667   \n",
      "44         middle                0.236055    0.402650         981   0.304791   \n",
      "45          early                0.050184    0.413725         510   0.303922   \n",
      "46           late                0.910655    0.398810         336   0.282738   \n",
      "47           late                0.959441    0.416409        1938   0.288958   \n",
      "48         middle                0.311841    0.414100        1461   0.321697   \n",
      "49          early                0.070215    0.384615         234   0.329060   \n",
      "50          early                0.039113    0.397163         282   0.312057   \n",
      "51           late                0.932583    0.394822         309   0.333333   \n",
      "52         middle                0.183952    0.389937         318   0.317610   \n",
      "53          early                0.093967    0.376068         117   0.290598   \n",
      "\n",
      "    T_Content  G_Content  C_Content  ...  kmer_CTG  kmer_CTC  kmer_CGA  \\\n",
      "0    0.267677   0.176768   0.166667  ...  0.005102  0.005102  0.005102   \n",
      "1    0.269406   0.216895   0.166667  ...  0.006881  0.004587  0.002294   \n",
      "2    0.277778   0.228889   0.164444  ...  0.013393  0.000000  0.006696   \n",
      "3    0.286448   0.218369   0.171484  ...  0.009003  0.007074  0.010289   \n",
      "4    0.299835   0.214227   0.177006  ...  0.013245  0.004967  0.008278   \n",
      "5    0.298851   0.258621   0.206897  ...  0.023256  0.000000  0.005814   \n",
      "6    0.295543   0.218023   0.202519  ...  0.022330  0.009709  0.005825   \n",
      "7    0.287189   0.227815   0.182962  ...  0.013562  0.006458  0.006458   \n",
      "8    0.302083   0.270833   0.140625  ...  0.010526  0.000000  0.010526   \n",
      "9    0.252525   0.205387   0.184063  ...  0.008999  0.005624  0.002250   \n",
      "10   0.314750   0.221835   0.153310  ...  0.015134  0.005821  0.011641   \n",
      "11   0.281879   0.203579   0.154362  ...  0.013483  0.000000  0.008989   \n",
      "12   0.264583   0.212500   0.189583  ...  0.002092  0.012552  0.004184   \n",
      "13   0.296482   0.226131   0.175879  ...  0.011765  0.003361  0.015126   \n",
      "14   0.276190   0.300000   0.128571  ...  0.000000  0.009615  0.000000   \n",
      "15   0.310292   0.211982   0.153610  ...  0.010786  0.004622  0.003082   \n",
      "16   0.254902   0.235294   0.147059  ...  0.014851  0.004950  0.000000   \n",
      "17   0.294686   0.183575   0.154589  ...  0.009756  0.004878  0.009756   \n",
      "18   0.316456   0.223629   0.143460  ...  0.012766  0.004255  0.004255   \n",
      "19   0.291667   0.216667   0.158333  ...  0.008475  0.000000  0.016949   \n",
      "20   0.302057   0.199657   0.170094  ...  0.012864  0.008148  0.006003   \n",
      "21   0.240363   0.244898   0.176871  ...  0.002278  0.006834  0.015945   \n",
      "22   0.292237   0.189498   0.198630  ...  0.022936  0.006881  0.011468   \n",
      "23   0.281705   0.225918   0.169439  ...  0.012483  0.004161  0.005548   \n",
      "24   0.290727   0.255639   0.167920  ...  0.020151  0.007557  0.002519   \n",
      "25   0.285604   0.237875   0.171671  ...  0.020817  0.002313  0.011565   \n",
      "26   0.272109   0.207483   0.183673  ...  0.020548  0.000000  0.010274   \n",
      "27   0.280042   0.219436   0.165099  ...  0.009424  0.005236  0.014660   \n",
      "28   0.250000   0.250000   0.142857  ...  0.000000  0.006024  0.000000   \n",
      "29   0.230159   0.285714   0.162698  ...  0.004000  0.016000  0.008000   \n",
      "30   0.300613   0.206544   0.179959  ...  0.020534  0.002053  0.004107   \n",
      "31   0.304813   0.199643   0.174688  ...  0.010733  0.005367  0.008945   \n",
      "32   0.237037   0.192593   0.133333  ...  0.007519  0.007519  0.000000   \n",
      "33   0.250871   0.222997   0.193961  ...  0.011641  0.006985  0.003492   \n",
      "34   0.292237   0.237443   0.109589  ...  0.009217  0.000000  0.000000   \n",
      "35   0.311955   0.210297   0.170157  ...  0.010480  0.005677  0.003057   \n",
      "36   0.304659   0.224014   0.164875  ...  0.016187  0.005396  0.010791   \n",
      "37   0.294974   0.226190   0.165344  ...  0.011936  0.007958  0.009284   \n",
      "38   0.292237   0.239726   0.141553  ...  0.011468  0.006881  0.009174   \n",
      "39   0.275298   0.248512   0.172619  ...  0.008955  0.007463  0.007463   \n",
      "40   0.220430   0.231183   0.209677  ...  0.016304  0.005435  0.005435   \n",
      "41   0.291005   0.222222   0.190476  ...  0.016043  0.005348  0.000000   \n",
      "42   0.303419   0.224359   0.163462  ...  0.008565  0.004283  0.009636   \n",
      "43   0.298667   0.234667   0.168000  ...  0.021448  0.005362  0.002681   \n",
      "44   0.292559   0.244648   0.158002  ...  0.014300  0.003064  0.005107   \n",
      "45   0.282353   0.247059   0.166667  ...  0.015748  0.007874  0.013780   \n",
      "46   0.318452   0.226190   0.172619  ...  0.026946  0.005988  0.005988   \n",
      "47   0.294634   0.238906   0.177503  ...  0.017045  0.010847  0.005165   \n",
      "48   0.264203   0.220397   0.193703  ...  0.009596  0.008225  0.010966   \n",
      "49   0.286325   0.192308   0.192308  ...  0.008621  0.012931  0.012931   \n",
      "50   0.290780   0.230496   0.166667  ...  0.025000  0.000000  0.014286   \n",
      "51   0.271845   0.229773   0.165049  ...  0.016287  0.000000  0.013029   \n",
      "52   0.292453   0.226415   0.163522  ...  0.012658  0.009494  0.003165   \n",
      "53   0.333333   0.222222   0.153846  ...  0.026087  0.000000  0.000000   \n",
      "\n",
      "    kmer_CGT  kmer_CGG  kmer_CGC  kmer_CCA  kmer_CCT  kmer_CCG  kmer_CCC  \n",
      "0   0.010204  0.000000  0.000000  0.015306  0.010204  0.005102  0.000000  \n",
      "1   0.013761  0.004587  0.018349  0.009174  0.002294  0.004587  0.000000  \n",
      "2   0.013393  0.006696  0.006696  0.006696  0.004464  0.004464  0.004464  \n",
      "3   0.021222  0.006431  0.002572  0.005145  0.007074  0.002572  0.004502  \n",
      "4   0.012417  0.003725  0.004553  0.012003  0.009520  0.004139  0.006209  \n",
      "5   0.011628  0.000000  0.017442  0.011628  0.017442  0.017442  0.017442  \n",
      "6   0.025243  0.005825  0.003883  0.006796  0.008738  0.006796  0.005825  \n",
      "7   0.016790  0.003552  0.004843  0.007104  0.010978  0.003229  0.004521  \n",
      "8   0.015789  0.021053  0.005263  0.005263  0.005263  0.010526  0.000000  \n",
      "9   0.006749  0.006749  0.003375  0.016873  0.007874  0.002250  0.007874  \n",
      "10  0.011641  0.005821  0.001164  0.003492  0.010477  0.002328  0.002328  \n",
      "11  0.011236  0.004494  0.002247  0.004494  0.011236  0.004494  0.006742  \n",
      "12  0.016736  0.004184  0.006276  0.014644  0.008368  0.006276  0.000000  \n",
      "13  0.013445  0.005042  0.003361  0.005042  0.010084  0.001681  0.005042  \n",
      "14  0.009615  0.009615  0.004808  0.009615  0.004808  0.000000  0.000000  \n",
      "15  0.013867  0.001541  0.003082  0.004622  0.013867  0.006163  0.006163  \n",
      "16  0.004950  0.000000  0.009901  0.000000  0.000000  0.000000  0.000000  \n",
      "17  0.009756  0.000000  0.000000  0.014634  0.009756  0.000000  0.009756  \n",
      "18  0.008511  0.017021  0.000000  0.004255  0.008511  0.004255  0.000000  \n",
      "19  0.000000  0.000000  0.000000  0.016949  0.008475  0.000000  0.000000  \n",
      "20  0.003859  0.004288  0.006861  0.008576  0.008576  0.001286  0.006432  \n",
      "21  0.013667  0.000000  0.004556  0.009112  0.004556  0.009112  0.002278  \n",
      "22  0.009174  0.002294  0.006881  0.006881  0.018349  0.002294  0.013761  \n",
      "23  0.014216  0.002774  0.004161  0.006241  0.007282  0.003467  0.003467  \n",
      "24  0.012594  0.012594  0.002519  0.015113  0.005038  0.002519  0.002519  \n",
      "25  0.009252  0.003855  0.006939  0.008481  0.007710  0.002313  0.001542  \n",
      "26  0.006849  0.013699  0.003425  0.013699  0.000000  0.006849  0.000000  \n",
      "27  0.007330  0.005236  0.005236  0.008377  0.008377  0.004188  0.004188  \n",
      "28  0.048193  0.000000  0.006024  0.000000  0.000000  0.000000  0.000000  \n",
      "29  0.012000  0.016000  0.000000  0.008000  0.012000  0.000000  0.004000  \n",
      "30  0.012320  0.002053  0.010267  0.012320  0.014374  0.000000  0.004107  \n",
      "31  0.025045  0.003578  0.001789  0.007156  0.010733  0.003578  0.003578  \n",
      "32  0.015038  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
      "33  0.009313  0.010477  0.004657  0.016298  0.006985  0.004657  0.001164  \n",
      "34  0.013825  0.004608  0.000000  0.009217  0.004608  0.000000  0.000000  \n",
      "35  0.012227  0.004367  0.003057  0.010480  0.010480  0.004803  0.003493  \n",
      "36  0.010791  0.007194  0.005396  0.007194  0.012590  0.005396  0.003597  \n",
      "37  0.009284  0.003979  0.001326  0.005305  0.010610  0.001326  0.003979  \n",
      "38  0.020642  0.011468  0.002294  0.006881  0.006881  0.004587  0.002294  \n",
      "39  0.010448  0.007463  0.004478  0.004478  0.005970  0.002985  0.002985  \n",
      "40  0.021739  0.000000  0.005435  0.005435  0.000000  0.005435  0.010870  \n",
      "41  0.010695  0.000000  0.010695  0.005348  0.021390  0.000000  0.000000  \n",
      "42  0.017131  0.002141  0.002141  0.009636  0.005353  0.002141  0.002141  \n",
      "43  0.016086  0.002681  0.002681  0.002681  0.013405  0.005362  0.008043  \n",
      "44  0.013279  0.004086  0.005107  0.002043  0.012257  0.005107  0.001021  \n",
      "45  0.013780  0.001969  0.005906  0.005906  0.005906  0.003937  0.001969  \n",
      "46  0.008982  0.005988  0.008982  0.000000  0.002994  0.000000  0.000000  \n",
      "47  0.019112  0.005165  0.005682  0.006198  0.009814  0.004649  0.003099  \n",
      "48  0.012337  0.006169  0.007539  0.010281  0.009596  0.006854  0.003427  \n",
      "49  0.012931  0.004310  0.004310  0.004310  0.017241  0.004310  0.000000  \n",
      "50  0.010714  0.007143  0.003571  0.007143  0.010714  0.000000  0.000000  \n",
      "51  0.013029  0.003257  0.000000  0.006515  0.009772  0.006515  0.003257  \n",
      "52  0.009494  0.015823  0.006329  0.003165  0.015823  0.003165  0.003165  \n",
      "53  0.017391  0.000000  0.000000  0.000000  0.008696  0.000000  0.000000  \n",
      "\n",
      "[54 rows x 80 columns]\n",
      "Features table saved as ../data/dna_feature_table\n",
      "                    Geneid                                        DNASequence  \\\n",
      "0    gene-CPT_phageK_gp066  ATGTATCCTGAAATAGATGTGGAAGAATTAGCGTATAAGCTAAAAA...   \n",
      "1    gene-CPT_phageK_gp079  ATGAATGGTATTATTGTATTTTACAAAGAAGAAAATAAACATATTC...   \n",
      "2    gene-CPT_phageK_gp220  ATGGATGCATCTATTCCTGAAGTACTATTAATTCTTAATTCAGCTA...   \n",
      "3    gene-CPT_phageK_gp161  ATGTCACAAGATAAATTAAGAGCAATTTACACAGAAATGAAAGTAG...   \n",
      "4    gene-CPT_phageK_gp043  ATGACAAACAAAAATTACTTATACGAAGAAACTCACACAGTACAAG...   \n",
      "..                     ...                                                ...   \n",
      "232  gene-CPT_phageK_gp101  ATGAGAAATAAAAAATCATTTCAAGAGCAGTTAAATGACATGCGTA...   \n",
      "233  gene-CPT_phageK_gp074  ATGGAATATTTATTTTTATTTATAGGTATTGGCATGATAATTTGGG...   \n",
      "234  gene-CPT_phageK_gp051  ATGATTCATATATTTTTAACTGATAGTTATGATAATAAAGTTTTAA...   \n",
      "235  gene-CPT_phageK_gp032  GTGGAGAGAACATTAAACTTATATGATTCCAAAGGTAAATTATTAA...   \n",
      "236  gene-CPT_phageK_gp105  ATGTTAAAAGGATTTTCAGAACATGTAGACAAACCTACAACTATTA...   \n",
      "\n",
      "    classification  Gene_Position_Relative  GC_Content  Seq_length  A_Content  \\\n",
      "0            early                0.902243    0.271845         309   0.449838   \n",
      "1            early                0.872698    0.281046         306   0.405229   \n",
      "2            early                0.138851    0.340171         585   0.364103   \n",
      "3            early                0.402833    0.259615         312   0.467949   \n",
      "4            early                0.972909    0.347222         288   0.388889   \n",
      "..             ...                     ...         ...         ...        ...   \n",
      "232         middle                0.819710    0.306383         705   0.394326   \n",
      "233          early                0.884410    0.277056         231   0.316017   \n",
      "234          early                0.951118    0.281250         288   0.364583   \n",
      "235           late                0.090583    0.283019         318   0.421384   \n",
      "236         middle                0.800245    0.302083         864   0.378472   \n",
      "\n",
      "     T_Content  G_Content  C_Content  ...  kmer_CTG  kmer_CTC  kmer_CGA  \\\n",
      "0     0.278317   0.187702   0.084142  ...  0.006515  0.000000  0.000000   \n",
      "1     0.313725   0.173203   0.107843  ...  0.000000  0.003289  0.000000   \n",
      "2     0.295726   0.200000   0.140171  ...  0.006861  0.006861  0.001715   \n",
      "3     0.272436   0.147436   0.112179  ...  0.003226  0.000000  0.000000   \n",
      "4     0.263889   0.194444   0.152778  ...  0.000000  0.003497  0.003497   \n",
      "..         ...        ...        ...  ...       ...       ...       ...   \n",
      "232   0.299291   0.180142   0.126241  ...  0.012802  0.009957  0.005690   \n",
      "233   0.406926   0.190476   0.086580  ...  0.000000  0.004367  0.000000   \n",
      "234   0.354167   0.215278   0.065972  ...  0.010490  0.006993  0.000000   \n",
      "235   0.295597   0.154088   0.128931  ...  0.003165  0.003165  0.006329   \n",
      "236   0.319444   0.173611   0.128472  ...  0.006961  0.006961  0.002320   \n",
      "\n",
      "     kmer_CGT  kmer_CGG  kmer_CGC  kmer_CCA  kmer_CCT  kmer_CCG  kmer_CCC  \n",
      "0    0.003257  0.003257  0.000000  0.003257  0.003257  0.000000  0.000000  \n",
      "1    0.003289  0.003289  0.000000  0.000000  0.009868  0.000000  0.000000  \n",
      "2    0.003431  0.001715  0.001715  0.001715  0.012007  0.003431  0.001715  \n",
      "3    0.003226  0.000000  0.000000  0.000000  0.006452  0.000000  0.000000  \n",
      "4    0.006993  0.020979  0.003497  0.013986  0.000000  0.000000  0.003497  \n",
      "..        ...       ...       ...       ...       ...       ...       ...  \n",
      "232  0.004267  0.001422  0.000000  0.005690  0.012802  0.002845  0.000000  \n",
      "233  0.000000  0.000000  0.000000  0.000000  0.008734  0.000000  0.000000  \n",
      "234  0.000000  0.003497  0.000000  0.000000  0.003497  0.000000  0.000000  \n",
      "235  0.003165  0.000000  0.003165  0.012658  0.000000  0.009494  0.003165  \n",
      "236  0.001160  0.001160  0.000000  0.002320  0.008121  0.000000  0.001160  \n",
      "\n",
      "[237 rows x 80 columns]\n"
     ]
    }
   ],
   "source": [
    "# Load feature_extraction table\n",
    "input_path = Path(\"../data/feature_extraction\")\n",
    "output_path = Path(\"../data/dna_feature_table\")\n",
    "gff_folder = Path(\"../data/features\")\n",
    "output_path.mkdir(exist_ok=True)\n",
    "\n",
    "# Computing k-mers + Counting k-mers for each sequence\n",
    "k = 3\n",
    "\n",
    "# Creates every possible k-mer of the length 3\n",
    "alphabet = [\"A\", \"T\", \"G\", \"C\"]\n",
    "all_kmers = [\"\".join(p) for p in product(alphabet, repeat=k)]\n",
    "top_k_features = 10\n",
    "\n",
    "\n",
    "# Helper function to count k-mers\n",
    "def count_kmers(seq, k):\n",
    "    seq = seq.upper()\n",
    "    counts = Counter(\n",
    "        seq[i : i + k]\n",
    "        for i in range(len(seq) - k + 1)\n",
    "        if set(seq[i : i + k]).issubset(alphabet)\n",
    "    )\n",
    "    return [\n",
    "        counts.get(kmer, 0) for kmer in all_kmers\n",
    "    ]  # returns a list with amount of each k-mer of a sequence\n",
    "\n",
    "\n",
    "for file in input_path.glob(\"*.tsv\"):\n",
    "    df = pd.read_csv(file, sep=\"\\t\")\n",
    "\n",
    "    # Only keep relevant columns\n",
    "    df_features = df[[\"Geneid\", \"DNASequence\", \"classification\"]].copy()\n",
    "\n",
    "    # Extracting Gene positions from GFF\n",
    "    base_name = file.name.split(\"_\")[0].lower()\n",
    "    gff_match = None\n",
    "    for gff_file in gff_folder.glob(\"*.gff3\"):\n",
    "        if gff_file.name.lower().startswith(base_name):\n",
    "            gff_match = gff_file\n",
    "            break\n",
    "    if gff_match:\n",
    "        gene_positions = {}\n",
    "        # Parse the GFF file to extract gene positions\n",
    "        with open(gff_match) as in_handle:\n",
    "            for rec in GFF.parse(in_handle):\n",
    "                genome_length = len(rec)\n",
    "                for feature in rec.features:\n",
    "                    if feature.type == \"gene\":\n",
    "                        gene_id = feature.id\n",
    "                        start = int(feature.location.start)\n",
    "                        end = int(feature.location.end)\n",
    "                        midpoint = (start + end) // 2\n",
    "                        gene_positions[gene_id] = (start, end, midpoint)\n",
    "        # Mapping the gene positions to the feature table\n",
    "        df_features[\"Gene_Position_Start\"] = df_features[\"Geneid\"].map(\n",
    "            lambda gid: gene_positions.get(gid, (None, None, None))[0]\n",
    "        )\n",
    "        df_features[\"Gene_Position_End\"] = df_features[\"Geneid\"].map(\n",
    "            lambda gid: gene_positions.get(gid, (None, None, None))[1]\n",
    "        )\n",
    "        df_features[\"Gene_Position_Midpoint\"] = df_features[\"Geneid\"].map(\n",
    "            lambda gid: gene_positions.get(gid, (None, None, None))[2]\n",
    "        )\n",
    "        # Calculate the relative position of each gene ( midpoint / genome length)\n",
    "        df_features[\"Gene_Position_Relative\"] = df_features[\n",
    "            \"Gene_Position_Midpoint\"\n",
    "        ].apply(lambda x: x / genome_length if x is not None else None)\n",
    "        # Remove Start, End and Midpoint columns\n",
    "        df_features.drop(\n",
    "            [\"Gene_Position_Start\", \"Gene_Position_End\", \"Gene_Position_Midpoint\"],\n",
    "            axis=1,\n",
    "            inplace=True,\n",
    "        )\n",
    "    else:\n",
    "        print(f\"No GFF found for {file.name}, skipping position features.\")\n",
    "\n",
    "    # Computing GC-content for each sequence\n",
    "    df_features[\"GC_Content\"] = df_features[\"DNASequence\"].apply(\n",
    "        lambda seq: gc_fraction(Seq(seq))\n",
    "    )\n",
    "\n",
    "    # Computing sequence length for each DNA sequence\n",
    "    df_features[\"Seq_length\"] = df_features[\"DNASequence\"].apply(\n",
    "        lambda seq: len(seq) if isinstance(seq, str) else 0\n",
    "    )\n",
    "\n",
    "    # Function to count bases\n",
    "    def count_bases(seq):\n",
    "        seq = seq.upper()\n",
    "        return {\n",
    "            \"A_Content\": seq.count(\"A\"),\n",
    "            \"T_Content\": seq.count(\"T\"),\n",
    "            \"G_Content\": seq.count(\"G\"),\n",
    "            \"C_Content\": seq.count(\"C\"),\n",
    "        }\n",
    "\n",
    "    motives = {\n",
    "        \"TATA_box\": \"TATAAA\",  # Promotor\n",
    "        \"CAAT_box\": \"GGCCAATCT\",  # Enhancer\n",
    "        \"PolyA_signal\": \"AATAAA\",  # Signal for transcriptation\n",
    "    }\n",
    "\n",
    "    def count_motives(sequence, motives):\n",
    "        sequence = sequence.upper()\n",
    "        result = {}\n",
    "        for name, pattern in motives.items():\n",
    "            result[f\"motive_{name}\"] = len(re.findall(pattern, sequence))\n",
    "        return result\n",
    "\n",
    "    # Counting bases A,G,C and T\n",
    "    base_counts = df_features[\"DNASequence\"].apply(count_bases).apply(pd.Series)\n",
    "    base_fractions = base_counts.div(df_features[\"Seq_length\"], axis=0)\n",
    "    df_features = pd.concat([df_features, base_fractions], axis=1)\n",
    "\n",
    "    # Counting Purin (A + G) and Pyrimidin (C + T)\n",
    "    df_features[\"Purin_Content\"] = df_features[\"DNASequence\"].apply(\n",
    "        lambda seq: (seq.count(\"A\") + seq.count(\"G\")) / len(seq)\n",
    "    )\n",
    "    df_features[\"Pyrimidin_Content\"] = df_features[\"DNASequence\"].apply(\n",
    "        lambda seq: (seq.count(\"C\") + seq.count(\"T\")) / len(seq)\n",
    "    )\n",
    "\n",
    "    # Computing CpG bias\n",
    "    def calc_cpg_bias(row):\n",
    "        # Annahme: row['DNASequence'], row['G-content'], row['C-content'], row['Seq_length'] existieren\n",
    "        seq = row[\"DNASequence\"].upper()\n",
    "        c_count = seq.count(\"C\")\n",
    "        g_count = seq.count(\"G\")\n",
    "        seq_length = len(seq)\n",
    "        cpg_count = sum(1 for i in range(seq_length - 1) if seq[i : i + 2] == \"CG\")\n",
    "        expected = (c_count * g_count) / seq_length if seq_length > 0 else 0\n",
    "        return (cpg_count / expected) if expected > 0 else 0\n",
    "\n",
    "    df_features[\"CpG_bias\"] = df_features.apply(calc_cpg_bias, axis=1)\n",
    "\n",
    "    # Counting motives\n",
    "    motive_features_rel = df_features.apply(\n",
    "        lambda row: pd.Series(\n",
    "            {\n",
    "                f\"{k}_rel\": v / row[\"Seq_length\"] if row[\"Seq_length\"] > 0 else 0\n",
    "                for k, v in count_motives(row[\"DNASequence\"], motives).items()\n",
    "            }\n",
    "        ),\n",
    "        axis=1,\n",
    "    )\n",
    "    df_features = pd.concat([df_features, motive_features_rel], axis=1)\n",
    "\n",
    "    # k-mer counting\n",
    "    kmer_counts = df_features[\"DNASequence\"].apply(lambda seq: count_kmers(seq, k))\n",
    "    kmer_df = pd.DataFrame(\n",
    "        kmer_counts.tolist(), columns=[f\"kmer_{kmer}\" for kmer in all_kmers]\n",
    "    )\n",
    "\n",
    "    # Normalize k-mer counts\n",
    "    kmer_sums = kmer_df.sum(axis=1).replace(0, 1)\n",
    "    kmer_df_norm = kmer_df.div(kmer_sums, axis=0)\n",
    "\n",
    "    # Save as new TSV\n",
    "    df_out = pd.concat([df_features, kmer_df_norm], axis=1)\n",
    "    out_file = output_path / file.name\n",
    "    df_out.to_csv(out_file, sep=\"\\t\", index=False)\n",
    "    print(f\"Features table saved as {output_path}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5db4f3c",
   "metadata": {},
   "source": [
    "# User-story 12: Protein based features\n",
    "@dottting  \n",
    "@davmar01\n",
    "\n",
    "The following code computes feature vectors for each sequence using amino acid composition (AAC) and dipeptide (k=2) frequencies, capturing both global and local sequence properties. The feature matrices are save at \"../features/protein_primary_k=1and2\". "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4f707b38",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved: ../features/protein/protein_primary_k=1_k=2\\Brandao_MCCM_full_raw_counts_tpm_filtered_classified_features.tsv\n",
      "Saved: ../features/protein/protein_primary_k=1_k=2\\Finstrlova_Newman_full_raw_counts_tpm_filtered_classified_features.tsv\n",
      "Saved: ../features/protein/protein_primary_k=1_k=2\\Guegler_T4_minusToxIN_full_raw_counts_tpm_filtered_classified_features.tsv\n",
      "Saved: ../features/protein/protein_primary_k=1_k=2\\Guegler_T7_plusToxIN_full_raw_counts_tpm_filtered_classified_features.tsv\n",
      "Saved: ../features/protein/protein_primary_k=1_k=2\\Lood_full_raw_counts_tpm_filtered_classified_features.tsv\n",
      "Saved: ../features/protein/protein_primary_k=1_k=2\\Sprenger_VC_WT_VP882_delta_cpdS_full_raw_counts_tpm_filtered_classified_features.tsv\n",
      "Saved: ../features/protein/protein_primary_k=1_k=2\\Yang_full_raw_counts_tpm_filtered_classified_features.tsv\n"
     ]
    }
   ],
   "source": [
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "\n",
    "input_folder = \"../data/feature_extraction\"\n",
    "output_folder = \"../features/protein/protein_primary_k=1_k=2\"\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "# List of standard amino acids\n",
    "AA_LIST = list(\"ACDEFGHIKLMNPQRSTVWY\")\n",
    "\n",
    "# Generate all possible 2-mer combinations (dipeptides)\n",
    "DIPEPTIDES = [\"\".join(p) for p in product(AA_LIST, repeat=2)]\n",
    "\n",
    "\n",
    "def extract_aac_kmer_features(seq):\n",
    "    \"\"\"Extract AAC, dipeptide (k=2), and key physicochemical features from a protein sequence.\"\"\"\n",
    "    seq = seq.upper().replace(\"*\", \"\")\n",
    "    total_len = len(seq)\n",
    "\n",
    "    # AAC features\n",
    "    aac_counts = Counter(seq)\n",
    "    aac_features = {f\"AAC_{aa}\": aac_counts.get(aa, 0) / total_len for aa in AA_LIST}\n",
    "\n",
    "    # Dipeptide (k=2) features\n",
    "    kmer_counts = Counter(seq[i : i + 2] for i in range(len(seq) - 1))\n",
    "    total_kmers = sum(kmer_counts.values()) or 1  # avoid division by zero\n",
    "    kmer_features = {\n",
    "        f\"kmer_{dp}\": kmer_counts.get(dp, 0) / total_kmers for dp in DIPEPTIDES\n",
    "    }\n",
    "\n",
    "    # Physicochemical features\n",
    "    try:\n",
    "        analysis = ProteinAnalysis(seq)\n",
    "        pI = analysis.isoelectric_point()\n",
    "        gravy = analysis.gravy()\n",
    "        instability = analysis.instability_index()\n",
    "    except Exception:\n",
    "        pI, gravy, instability = None, None, None\n",
    "\n",
    "    extra_features = {\n",
    "        \"IsoelectricPoint\": pI,\n",
    "        \"GRAVY\": gravy,\n",
    "        \"InstabilityIndex\": instability,\n",
    "        \"seq_length\": total_len,\n",
    "    }\n",
    "\n",
    "    return {**aac_features, **kmer_features, **extra_features}\n",
    "\n",
    "\n",
    "# Process each .tsv file in the input directory\n",
    "for filename in os.listdir(input_folder):\n",
    "    if not filename.endswith(\".tsv\"):\n",
    "        continue\n",
    "\n",
    "    filepath = os.path.join(input_folder, filename)\n",
    "\n",
    "    try:\n",
    "        df = pd.read_csv(filepath, sep=\"\\t\")\n",
    "        df.columns = df.columns.str.strip().str.lower()\n",
    "    except Exception as e:\n",
    "        print(f\"Error reading {filename}: {e}\")\n",
    "        continue\n",
    "\n",
    "    # Check for required columns\n",
    "    required = {\"geneid\", \"classification\", \"proteinsequence\"}\n",
    "    if not required.issubset(df.columns):\n",
    "        print(f\"Skipping file {filename} â€“ required columns missing.\")\n",
    "        print(f\"Found columns: {df.columns.tolist()}\")\n",
    "        continue\n",
    "\n",
    "    # Extract features row-wise\n",
    "    all_features = []\n",
    "    for _, row in df.iterrows():\n",
    "        try:\n",
    "            feats = extract_aac_kmer_features(row[\"proteinsequence\"])\n",
    "            ordered_feats = {\n",
    "                \"Geneid\": row[\"geneid\"],\n",
    "                \"classification\": row[\"classification\"],\n",
    "                **feats,\n",
    "            }\n",
    "            all_features.append(ordered_feats)\n",
    "        except Exception as e:\n",
    "            print(f\"Error in row with gene ID {row.get('geneid', 'unknown')}: {e}\")\n",
    "\n",
    "    # Save the resulting feature matrix\n",
    "    if all_features:\n",
    "        output_df = pd.DataFrame(all_features)\n",
    "        output_filename = filename.replace(\".tsv\", \"_features.tsv\")\n",
    "        output_path = os.path.join(output_folder, output_filename)\n",
    "        output_df.to_csv(output_path, sep=\"\\t\", index=False)\n",
    "        print(f\"Saved: {output_path}\")\n",
    "    else:\n",
    "        print(f\"No valid entries in {filename}, nothing saved.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49a02e2d",
   "metadata": {},
   "source": [
    "### Tripedtide features (most commonly used)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ad09012",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved k=3+physicochemical features to: ../features/protein/protein_primary_table_k=3\\Brandao_MCCM_full_raw_counts_tpm_filtered_classified_k3_physchem.tsv\n",
      "Saved k=3+physicochemical features to: ../features/protein/protein_primary_table_k=3\\Finstrlova_Newman_full_raw_counts_tpm_filtered_classified_k3_physchem.tsv\n",
      "Saved k=3+physicochemical features to: ../features/protein/protein_primary_table_k=3\\Guegler_T4_minusToxIN_full_raw_counts_tpm_filtered_classified_k3_physchem.tsv\n",
      "Saved k=3+physicochemical features to: ../features/protein/protein_primary_table_k=3\\Guegler_T7_plusToxIN_full_raw_counts_tpm_filtered_classified_k3_physchem.tsv\n",
      "Saved k=3+physicochemical features to: ../features/protein/protein_primary_table_k=3\\Lood_full_raw_counts_tpm_filtered_classified_k3_physchem.tsv\n",
      "Saved k=3+physicochemical features to: ../features/protein/protein_primary_table_k=3\\Sprenger_VC_WT_VP882_delta_cpdS_full_raw_counts_tpm_filtered_classified_k3_physchem.tsv\n",
      "Saved k=3+physicochemical features to: ../features/protein/protein_primary_table_k=3\\Yang_full_raw_counts_tpm_filtered_classified_k3_physchem.tsv\n"
     ]
    }
   ],
   "source": [
    "output_folder = \"../features/protein/protein_primary_table_k=3\"\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "TRIPEPTIDES = [\"\".join(p) for p in product(AA_LIST, repeat=3)]\n",
    "\n",
    "\n",
    "def extract_k3_features(seq):\n",
    "    \"\"\"Extract tripeptide (k=3) frequency features + top 3 physicochemical features from a protein sequence.\"\"\"\n",
    "    seq = seq.upper().replace(\"*\", \"\")\n",
    "\n",
    "    # Tripeptide features\n",
    "    tri_counts = Counter(seq[i : i + 3] for i in range(len(seq) - 2))\n",
    "    total_tri = sum(tri_counts.values()) or 1\n",
    "    tri_features = {\n",
    "        f\"kmer3_{tp}\": tri_counts.get(tp, 0) / total_tri for tp in TRIPEPTIDES\n",
    "    }\n",
    "\n",
    "    # Physicochemical features\n",
    "    try:\n",
    "        analysis = ProteinAnalysis(seq)\n",
    "        pI = analysis.isoelectric_point()\n",
    "        gravy = analysis.gravy()\n",
    "        instability = analysis.instability_index()\n",
    "    except Exception:\n",
    "        pI, gravy, instability = None, None, None\n",
    "\n",
    "    extra_features = {\n",
    "        \"IsoelectricPoint\": pI,\n",
    "        \"GRAVY\": gravy,\n",
    "        \"InstabilityIndex\": instability,\n",
    "        \"seq_length\": len(seq),\n",
    "    }\n",
    "\n",
    "    return {**tri_features, **extra_features}\n",
    "\n",
    "\n",
    "# Process each .tsv file in the input directory\n",
    "for filename in os.listdir(input_folder):\n",
    "    if not filename.endswith(\".tsv\"):\n",
    "        continue\n",
    "\n",
    "    filepath = os.path.join(input_folder, filename)\n",
    "\n",
    "    try:\n",
    "        df = pd.read_csv(filepath, sep=\"\\t\")\n",
    "        df.columns = df.columns.str.strip().str.lower()\n",
    "    except Exception as e:\n",
    "        print(f\"Error reading {filename}: {e}\")\n",
    "        continue\n",
    "\n",
    "    # Check for required columns\n",
    "    required = {\"geneid\", \"classification\", \"proteinsequence\"}\n",
    "    if not required.issubset(df.columns):\n",
    "        print(f\"Skipping file {filename} â€“ required columns missing.\")\n",
    "        print(f\"Found columns: {df.columns.tolist()}\")\n",
    "        continue\n",
    "\n",
    "    # Extract features row-wise\n",
    "    all_features = []\n",
    "    for _, row in df.iterrows():\n",
    "        try:\n",
    "            feats = extract_k3_features(row[\"proteinsequence\"])\n",
    "            ordered_feats = {\n",
    "                \"Geneid\": row[\"geneid\"],\n",
    "                \"classification\": row[\"classification\"],\n",
    "                **feats,\n",
    "            }\n",
    "            all_features.append(ordered_feats)\n",
    "        except Exception as e:\n",
    "            print(f\"Error in row with gene ID {row.get('geneid', 'unknown')}: {e}\")\n",
    "\n",
    "    # Save the resulting feature matrix\n",
    "    if all_features:\n",
    "        output_df = pd.DataFrame(all_features)\n",
    "        output_filename = filename.replace(\".tsv\", \"_k3_physchem.tsv\")\n",
    "        output_path = os.path.join(output_folder, output_filename)\n",
    "        output_df.to_csv(output_path, sep=\"\\t\", index=False)\n",
    "        print(f\"Saved k=3+physicochemical features to: {output_path}\")\n",
    "    else:\n",
    "        print(f\"No valid entries in {filename}, nothing saved.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "001e7312",
   "metadata": {},
   "source": [
    "### Secondary Structure Features\n",
    "\n",
    "- Predict structures with ESM3  (see `data/protein_structures/pdb`)\n",
    "\n",
    "- Secondary-structure related\n",
    "  - Secondary-structure content (Î±-helices, Î²-sheets, coils)  \n",
    "  - Solvent accessibility  \n",
    "  - Secondary-structure transitions  \n",
    "\n",
    "- ESM3 embeddings (TODO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3af1a88e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mmcif files generated outside of this notebook\n",
    "from DSSPparser import parseDSSP\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Max ASA values per residue for RSA (relative solvent accessibility) (Tien et al., 2013 (emp.))\n",
    "MAX_ASA = {\n",
    "    \"A\": 121.0,\n",
    "    \"R\": 265.0,\n",
    "    \"N\": 187.0,\n",
    "    \"D\": 187.0,\n",
    "    \"C\": 148.0,\n",
    "    \"Q\": 214.0,\n",
    "    \"E\": 214.0,\n",
    "    \"G\": 97.0,\n",
    "    \"H\": 216.0,\n",
    "    \"I\": 195.0,\n",
    "    \"L\": 191.0,\n",
    "    \"K\": 230.0,\n",
    "    \"M\": 203.0,\n",
    "    \"F\": 228.0,\n",
    "    \"P\": 154.0,\n",
    "    \"S\": 143.0,\n",
    "    \"T\": 163.0,\n",
    "    \"W\": 264.0,\n",
    "    \"Y\": 255.0,\n",
    "    \"V\": 165.0,\n",
    "}\n",
    "\n",
    "\n",
    "def dssp_features(dssp_file):\n",
    "    features = {}\n",
    "\n",
    "    # Parse dssp file\n",
    "    parser = parseDSSP(dssp_file)\n",
    "    parser.parse()\n",
    "    df_dssp = parser.dictTodataframe()\n",
    "    # replace \" \" with \"C\" and strip whitespaces in struc column (represents coil)\n",
    "    df_dssp[\"struct\"] = df_dssp[\"struct\"].apply(\n",
    "        lambda x: \"C\" if x == \"   \" else x.strip()\n",
    "    )\n",
    "    # Normalize helix types\n",
    "    df_dssp[\"struct\"] = df_dssp[\"struct\"].apply(\n",
    "        lambda x: \"H\" if x in [\"G\", \"I\"] else x.strip()\n",
    "    )\n",
    "\n",
    "    ## Secondary structure content\n",
    "    # dssp: H, G, I = helix, E = beta strand, C = coil/loop\n",
    "    # The rest (T, S) are ignored\n",
    "    structure_counts = df_dssp[\"struct\"].value_counts()\n",
    "    total = len(df_dssp)\n",
    "    struc_content = {k: v / total for k, v in structure_counts.items()}\n",
    "    # Set default values (0.0)\n",
    "    for frac in [\"H\", \"E\", \"C\"]:\n",
    "        struc_content.setdefault(frac, 0.0)\n",
    "    # print(struc_content)\n",
    "\n",
    "    features = {f\"{k}_frac\": struc_content[k] for k in [\"H\", \"E\", \"C\"]}\n",
    "\n",
    "    ## Solvent Accessibility\n",
    "    # Calculate RSA\n",
    "    df_dssp[\"acc\"] = pd.to_numeric(df_dssp[\"acc\"])\n",
    "    df_dssp[\"rsa\"] = df_dssp.apply(\n",
    "        lambda row: row[\"acc\"] / MAX_ASA.get(row[\"aa\"], np.nan), axis=1\n",
    "    )\n",
    "    # Get mean RSA\n",
    "    features[\"rsa_mean\"] = df_dssp[\"rsa\"].mean().item()\n",
    "\n",
    "    # Fraction of exposed residues (threshold: RSA > 20%  (Chen, Zhou, Nucleic Acids Res, 2005))\n",
    "    features[\"frac_exposed\"] = df_dssp[\"rsa\"].apply(lambda x: x > 0.2).mean().item()\n",
    "\n",
    "    ## Secondary structure transitions per residue\n",
    "    # TODO: Verify viability of this: Count all transistions? combine all helix types etc and then count?\n",
    "    from collections import Counter\n",
    "\n",
    "    transitions = Counter()\n",
    "    structs = df_dssp[\"struct\"].tolist()\n",
    "    for i in range(1, len(structs)):\n",
    "        if not structs[i - 1] == structs[i]:\n",
    "            pair = (structs[i - 1], structs[i])\n",
    "            transitions[pair] += 1\n",
    "    features[\"transitions_per_residue\"] = sum(transitions.values()) / len(structs)\n",
    "\n",
    "    return features\n",
    "\n",
    "\n",
    "def protein_analysis_features(sequence):\n",
    "    if sequence == \"NOT_FOUND\" or \"X\" in sequence:\n",
    "        return\n",
    "    analysis = ProteinAnalysis(sequence)\n",
    "    return {\n",
    "        \"molecular_weight\": analysis.molecular_weight(),\n",
    "        \"aromaticity\": analysis.aromaticity(),\n",
    "    }\n",
    "\n",
    "\n",
    "# Import genes, onyl keep \"Geneid\" and \"Symbol\"\n",
    "tsv_path = Path(\"../data/feature_extraction/\").glob(\"*.tsv\")\n",
    "df_tsv = pd.DataFrame()\n",
    "for file in tsv_path:\n",
    "    df_temp = pd.read_csv(file, sep=\"\\t\")\n",
    "    df_temp = df_temp[[\"Geneid\", \"Symbol\", \"ProteinSequence\"]]\n",
    "    df_temp[\"Dataset\"] = file.name[:-4]\n",
    "    df_tsv = pd.concat([df_tsv, df_temp])\n",
    "df_tsv = df_tsv.reset_index(drop=True)\n",
    "\n",
    "# Import dssp files and run feature extraction\n",
    "dssp_files = Path(\"../data/protein_structures/dssp/\").glob(\"*.dssp\")\n",
    "df_features = pd.DataFrame()\n",
    "for file in dssp_files:\n",
    "    df_temp = pd.DataFrame([dssp_features(file)])\n",
    "    df_temp[\"Geneid\"] = file.name[:-5]\n",
    "    df_features = pd.concat([df_features, df_temp], ignore_index=True)\n",
    "\n",
    "# Other protein features\n",
    "df_tsv[\"properties\"] = df_tsv[\"ProteinSequence\"].apply(protein_analysis_features)\n",
    "# Expand dict into separate columns\n",
    "df_props = df_tsv[\"properties\"].apply(pd.Series)\n",
    "df_tsv = pd.concat([df_tsv, df_props], axis=1)\n",
    "df_tsv.drop(columns=[\"properties\", \"ProteinSequence\"], inplace=True)\n",
    "\n",
    "# Merge features and gene list and save\n",
    "df_tsv = df_tsv.merge(df_features, on=\"Geneid\", how=\"outer\")\n",
    "for dataset_name, df_group in df_tsv.groupby(\"Dataset\"):\n",
    "    df_group.drop(columns=[\"Symbol\", \"Dataset\"], inplace=True)\n",
    "    df_group.to_csv(\n",
    "        f\"../data/protein_structures/features/{dataset_name}_structure_features.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index=False,\n",
    "    )\n",
    "# df_tsv.drop(columns=[\"Symbol\", \"Dataset\"], inplace=True)\n",
    "# df_tsv.to_csv(\n",
    "#     \"../data/protein_structures/features/complete_structure_features.tsv\",\n",
    "#     sep=\"\\t\",\n",
    "# )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86513978",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge all feature tables\n",
    "\n",
    "# DNA Features data/dna_feature_table\n",
    "feature_list = [[], [], []]\n",
    "path = Path(\"../data/dna_feature_table\").glob(\"*.tsv\")\n",
    "for file in path:\n",
    "    feature_list[0].append(file)\n",
    "    pass\n",
    "\n",
    "# Protein seq features\n",
    "path = Path(\"../features/protein/protein_primary_table_k=3\").glob(\"*physchem.tsv\")\n",
    "for file in path:\n",
    "    feature_list[1].append(file)\n",
    "    pass\n",
    "\n",
    "# Protein structure features\n",
    "path = Path(\"../data/protein_structures/features\").glob(\"*.tsv\")\n",
    "for file in path:\n",
    "    feature_list[2].append(file)\n",
    "    pass\n",
    "\n",
    "for l1, l2, l3 in zip(\n",
    "    sorted(feature_list[0]), sorted(feature_list[1]), sorted(feature_list[2])\n",
    "):\n",
    "    df_dna = pd.read_csv(l1, sep=\"\\t\")\n",
    "    df_aa = pd.read_csv(l2, sep=\"\\t\")\n",
    "    df_struc = pd.read_csv(l3, sep=\"\\t\")\n",
    "\n",
    "    # df_complete = (\n",
    "    #     df_dna.merge(df_struc, on=\"Geneid\", how=\"outer\")\n",
    "    #     .merge(df_aa, on=\"Geneid\", how=\"outer\")\n",
    "    #     .drop(columns=\"classification_y\")\n",
    "    # )\n",
    "\n",
    "    df_complete = df_dna.merge(df_struc, on=\"Geneid\", how=\"outer\").rename(\n",
    "        columns={\"classification\": \"classification_x\"}\n",
    "    )\n",
    "\n",
    "    filename = l1.name[:-44] + \"_feature_table.tsv\"\n",
    "    df_complete.to_csv(f\"../data/feature_tables/{filename}\", sep=\"\\t\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "vividVirions",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
